From 1da64869162d528c60ab3d8020b0ae29fb54e53f Mon Sep 17 00:00:00 2001
From: Alec Ari <neotheuser@ymail.com>
Date: Wed, 26 Jun 2019 14:55:00 -0500
Subject: Countless Kconfig tweaks for IPIPE

Disables every problematic option I can think of and enables
several very important options for debugging. RTAI may cause
kernel panics on some systems so it is a good idea to always
have these on. BIGSMP and related options are also disabled.

Various other fixes and optimizations are included as well.

Make it easy to hack in native 32-bit support for testing.

Signed-off-by: Alec Ari <neotheuser@ymail.com>
---
 arch/Kconfig                        |   5 +-
 arch/x86/Kconfig                    |  62 +++++++-----
 arch/x86/Kconfig.debug              |  32 +++---
 arch/x86/kvm/Kconfig                |   1 +
 drivers/acpi/Kconfig                |  28 ++++--
 drivers/acpi/apei/Kconfig           |   2 +-
 drivers/acpi/dptf/Kconfig           |   2 +-
 drivers/acpi/nfit/Kconfig           |   1 +
 drivers/char/ipmi/Kconfig           |   2 +-
 drivers/char/tpm/Kconfig            |   2 +-
 drivers/cpufreq/Kconfig             |   1 +
 drivers/cpuidle/Kconfig             |   1 +
 drivers/crypto/Kconfig              |   1 +
 drivers/crypto/qat/Kconfig          |   6 +-
 drivers/devfreq/Kconfig             |   1 +
 drivers/firmware/Kconfig            |   2 +-
 drivers/fsi/Kconfig                 |   1 +
 drivers/gpu/drm/Kconfig             |   2 +-
 drivers/gpu/drm/gma500/Kconfig      |   2 +-
 drivers/gpu/drm/i915/Kconfig        |   2 +-
 drivers/gpu/drm/nouveau/Kconfig     |   2 +-
 drivers/gpu/drm/qxl/Kconfig         |   2 +-
 drivers/gpu/drm/vmwgfx/Kconfig      |   2 +-
 drivers/hwspinlock/Kconfig          |   1 +
 drivers/infiniband/hw/usnic/Kconfig |   1 +
 drivers/iommu/Kconfig               |   2 +-
 drivers/mailbox/Kconfig             |   1 +
 drivers/md/Kconfig                  |   2 +-
 drivers/misc/mic/Kconfig            |   6 +-
 drivers/misc/vmw_vmci/Kconfig       |   2 +-
 drivers/net/caif/Kconfig            |   3 +-
 drivers/pci/Kconfig                 |   9 +-
 drivers/pci/hotplug/Kconfig         |   2 +-
 drivers/pci/pcie/Kconfig            |   2 +-
 drivers/platform/x86/Kconfig        |   2 +-
 drivers/power/avs/Kconfig           |   1 +
 drivers/power/reset/Kconfig         |   1 +
 drivers/powercap/Kconfig            |   1 +
 drivers/pwm/Kconfig                 |   1 +
 drivers/regulator/Kconfig           |   1 +
 drivers/remoteproc/Kconfig          |   3 +-
 drivers/reset/Kconfig               |   2 +-
 drivers/rpmsg/Kconfig               |   2 +-
 drivers/scsi/ufs/Kconfig            |   2 +-
 drivers/thermal/Kconfig             |   1 +
 drivers/virt/Kconfig                |   1 +
 drivers/virtio/Kconfig              |   7 +-
 drivers/watchdog/Kconfig            |   1 +
 fs/Kconfig.binfmt                   |   1 +
 include/linux/ipipe.h               |   2 +-
 init/Kconfig                        |  33 +++---
 kernel/ipipe/Kconfig                |  19 ++--
 kernel/irq/Kconfig                  |   4 +-
 kernel/power/Kconfig                |   6 +-
 kernel/rcu/Kconfig.debug            |   8 +-
 kernel/trace/Kconfig                |   3 +-
 lib/Kconfig.debug                   | 150 ++++++++++++++++------------
 lib/Kconfig.kasan                   |   1 +
 lib/Kconfig.kgdb                    |   2 +-
 lib/Kconfig.ubsan                   |   1 +
 mm/Kconfig                          |   2 +-
 mm/Kconfig.debug                    |   6 +-
 samples/Kconfig                     |   1 +
 security/Kconfig                    |   6 +-
 64 files changed, 274 insertions(+), 190 deletions(-)

diff --git a/arch/Kconfig b/arch/Kconfig
index 77b3e21c4844..782589f1cb1f 100644
--- a/arch/Kconfig
+++ b/arch/Kconfig
@@ -402,7 +402,7 @@ config HAVE_GCC_PLUGINS
 menuconfig GCC_PLUGINS
 	bool "GCC plugins"
 	depends on HAVE_GCC_PLUGINS
-	depends on !COMPILE_TEST
+	depends on !COMPILE_TEST && !IPIPE
 	help
 	  GCC plugins are loadable modules that provide extra features to the
 	  compiler. They are useful for runtime instrumentation and static analysis.
@@ -570,6 +570,7 @@ config CC_STACKPROTECTOR_REGULAR
 
 config CC_STACKPROTECTOR_STRONG
 	bool "Strong"
+	depends on !IPIPE
 	select CC_STACKPROTECTOR
 	help
 	  Functions will have the stack-protector canary logic added in any
@@ -598,6 +599,7 @@ config THIN_ARCHIVES
 
 config LD_DEAD_CODE_DATA_ELIMINATION
 	bool
+	depends on !IPIPE
 	help
 	  Select this if the architecture wants to do dead code and
 	  data elimination with the linker by compiling with
@@ -958,6 +960,7 @@ config ARCH_HAS_REFCOUNT
 
 config REFCOUNT_FULL
 	bool "Perform full reference count validation at the expense of speed"
+	depends on !IPIPE
 	help
 	  Enabling this switches the refcounting infrastructure from a fast
 	  unchecked atomic_t implementation to a fully state checked
diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index b147b5c53af6..8f259be4671a 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -145,8 +145,9 @@ config X86
 	select HAVE_IOREMAP_PROT
 	select HAVE_IRQ_EXIT_ON_IRQ_STACK	if X86_64
 	select HAVE_IRQ_TIME_ACCOUNTING
-	select HAVE_IPIPE_SUPPORT		if X86_64
 	select HAVE_IPIPE_TRACER_SUPPORT
+	# Disabling IPIPE is broken right now (Xenomai mainline)
+	select IPIPE				if X86_64
 	select IPIPE_HAVE_SAFE_THREAD_INFO if IPIPE
 	select IPIPE_WANT_PTE_PINNING if IPIPE
 	select IPIPE_HAVE_VM_NOTIFIER if IPIPE
@@ -427,7 +428,7 @@ config X86_MPPARSE
 
 config X86_BIGSMP
 	bool "Support for big SMP systems with more than 8 CPUs"
-	depends on X86_32 && SMP
+	depends on X86_32 && SMP && !IPIPE
 	---help---
 	  This option is needed for the systems that have more than 8 CPUs
 
@@ -438,6 +439,7 @@ config GOLDFISH
 config RETPOLINE
 	bool "Avoid speculative indirect branches in kernel"
 	default y
+	depends on !IPIPE
 	select STACK_VALIDATION if HAVE_STACK_VALIDATION
 	help
 	  Compile kernel with the retpoline compiler options to guard against
@@ -461,6 +463,7 @@ config INTEL_RDT
 if X86_32
 config X86_EXTENDED_PLATFORM
 	bool "Support for extended (non-PC) x86 platforms"
+	depends on !IPIPE
 	default y
 	---help---
 	  If you disable this option then the kernel will only support
@@ -483,6 +486,7 @@ endif
 if X86_64
 config X86_EXTENDED_PLATFORM
 	bool "Support for extended (non-PC) x86 platforms"
+	depends on !IPIPE
 	default y
 	---help---
 	  If you disable this option then the kernel will only support
@@ -603,7 +607,7 @@ config X86_INTEL_QUARK
 
 config X86_INTEL_LPSS
 	bool "Intel Low Power Subsystem Support"
-	depends on X86 && ACPI
+	depends on X86 && ACPI && !IPIPE
 	select COMMON_CLK
 	select PINCTRL
 	select IOSF_MBI
@@ -729,6 +733,7 @@ config SCHED_OMIT_FRAME_POINTER
 
 menuconfig HYPERVISOR_GUEST
 	bool "Linux guest support"
+	depends on !IPIPE
 	---help---
 	  Say Y here to enable options for running Linux under various hyper-
 	  visors. This option enables basic hypervisor detection and platform
@@ -943,6 +948,7 @@ config MAXSMP
 config NR_CPUS
 	int "Maximum number of CPUs" if SMP && !MAXSMP
 	range 2 8 if SMP && X86_32 && !X86_BIGSMP
+	range 2 16 if X86_64 && IPIPE
 	range 2 64 if SMP && X86_32 && X86_BIGSMP
 	range 2 512 if SMP && !MAXSMP && !CPUMASK_OFFSTACK && X86_64
 	range 2 8192 if SMP && !MAXSMP && CPUMASK_OFFSTACK && X86_64
@@ -957,16 +963,19 @@ config NR_CPUS
 	  supported value is 8192, otherwise the maximum value is 512.  The
 	  minimum value which makes sense is 2.
 
-	  This is purely to save memory - each supported CPU adds
-	  approximately eight kilobytes to the kernel image.
+	  For IPIPE-enabled platforms, large values causes serious problems.
+	  The max limit in this case is 16.
+
+	  Each supported CPU adds approximately 8KB to the kernel image.
 
 config SCHED_SMT
 	def_bool y if SMP
+	depends on !IPIPE
 
 config SCHED_MC
 	def_bool y
 	prompt "Multi-core scheduler support"
-	depends on SMP
+	depends on SMP && !IPIPE
 	---help---
 	  Multi-core scheduler support improves the CPU scheduler's decision
 	  making when dealing with multi-core CPU chips at a cost of slightly
@@ -1123,7 +1132,7 @@ source "arch/x86/events/Kconfig"
 config X86_LEGACY_VM86
 	bool "Legacy VM86 support"
 	default n
-	depends on X86_32
+	depends on X86_32 && !IPIPE
 	---help---
 	  This option allows user programs to put the CPU into V8086
 	  mode, which is an 80286-era approximation of 16-bit real mode.
@@ -1155,7 +1164,7 @@ config VM86
 config X86_16BIT
 	bool "Enable support for 16-bit segments" if EXPERT
 	default y
-	depends on MODIFY_LDT_SYSCALL
+	depends on MODIFY_LDT_SYSCALL && !IPIPE
 	---help---
 	  This option is required by programs like Wine to run 16-bit
 	  protected mode legacy code on x86 processors.  Disabling
@@ -1173,7 +1182,7 @@ config X86_ESPFIX64
 config X86_VSYSCALL_EMULATION
        bool "Enable vsyscall emulation" if EXPERT
        default y
-       depends on X86_64
+       depends on X86_64 && !IPIPE
        ---help---
 	 This enables emulation of the legacy vsyscall page.  Disabling
 	 it is roughly equivalent to booting with vsyscall=none, except
@@ -1190,7 +1199,7 @@ config X86_VSYSCALL_EMULATION
 
 config TOSHIBA
 	tristate "Toshiba Laptop support"
-	depends on X86_32
+	depends on X86_32 && !IPIPE
 	---help---
 	  This adds a driver to safely access the System Management Mode of
 	  the CPU on Toshiba portables with a genuine Toshiba BIOS. It does
@@ -1206,6 +1215,7 @@ config TOSHIBA
 
 config I8K
 	tristate "Dell i8k legacy laptop support"
+	depends on !IPIPE
 	select HWMON
 	select SENSORS_DELL_SMM
 	---help---
@@ -1484,6 +1494,7 @@ config NUMA
 	bool "Numa Memory Allocation and Scheduler Support"
 	depends on SMP
 	depends on X86_64 || (X86_32 && HIGHMEM64G && X86_BIGSMP)
+	depends on !IPIPE
 	default y if X86_BIGSMP
 	---help---
 	  Enable NUMA (Non Uniform Memory Access) support.
@@ -1605,8 +1616,7 @@ config X86_PMEM_LEGACY_DEVICE
 
 config X86_PMEM_LEGACY
 	tristate "Support non-standard NVDIMMs and ADR protected memory"
-	depends on PHYS_ADDR_T_64BIT
-	depends on BLK_DEV
+	depends on PHYS_ADDR_T_64BIT && BLK_DEV && !IPIPE
 	select X86_PMEM_LEGACY_DEVICE
 	select LIBNVDIMM
 	help
@@ -1808,6 +1818,7 @@ config ARCH_RANDOM
 
 config X86_SMAP
 	def_bool y
+	depends on !IPIPE
 	prompt "Supervisor Mode Access Prevention" if EXPERT
 	---help---
 	  Supervisor Mode Access Prevention (SMAP) is a security
@@ -1821,7 +1832,7 @@ config X86_INTEL_MPX
 	prompt "Intel MPX (Memory Protection Extensions)"
 	def_bool n
 	# Note: only available in 64-bit mode due to VMA flags shortage
-	depends on CPU_SUP_INTEL && X86_64
+	depends on CPU_SUP_INTEL && X86_64 && !IPIPE
 	select ARCH_USES_HIGH_VMA_FLAGS
 	---help---
 	  MPX provides hardware features that can be used in
@@ -1849,7 +1860,7 @@ config X86_INTEL_MEMORY_PROTECTION_KEYS
 	prompt "Intel Memory Protection Keys"
 	def_bool y
 	# Note: only available in 64-bit mode
-	depends on CPU_SUP_INTEL && X86_64
+	depends on CPU_SUP_INTEL && X86_64 && !IPIPE
 	select ARCH_USES_HIGH_VMA_FLAGS
 	select ARCH_HAS_PKEYS
 	---help---
@@ -1863,7 +1874,7 @@ config X86_INTEL_MEMORY_PROTECTION_KEYS
 
 config EFI
 	bool "EFI runtime service support"
-	depends on ACPI
+	depends on ACPI && !IPIPE
 	select UCS2_STRING
 	select EFI_RUNTIME_WRAPPERS
 	---help---
@@ -1921,6 +1932,7 @@ source kernel/Kconfig.hz
 
 config KEXEC
 	bool "kexec system call"
+	depends on !IPIPE
 	select KEXEC_CORE
 	---help---
 	  kexec is a system call that implements the ability to shutdown your
@@ -1940,9 +1952,8 @@ config KEXEC_FILE
 	bool "kexec file based system call"
 	select KEXEC_CORE
 	select BUILD_BIN2C
-	depends on X86_64
-	depends on CRYPTO=y
-	depends on CRYPTO_SHA256=y
+	depends on X86_64 && CRYPTO=y && CRYPTO_SHA256=y
+	depends on !IPIPE
 	---help---
 	  This is new version of kexec system call. This system call is
 	  file based and takes file descriptors as system call argument
@@ -1971,6 +1982,7 @@ config KEXEC_BZIMAGE_VERIFY_SIG
 config CRASH_DUMP
 	bool "kernel crash dumps"
 	depends on X86_64 || (X86_32 && HIGHMEM)
+	depends on !IPIPE
 	---help---
 	  Generate crash dump after being started by kexec.
 	  This should be normally only set in special crash dump kernels
@@ -2032,6 +2044,7 @@ config PHYSICAL_START
 
 config RELOCATABLE
 	bool "Build a relocatable kernel"
+	depends on !IPIPE
 	default y
 	---help---
 	  This builds a kernel image that retains relocation information
@@ -2148,7 +2161,7 @@ config RANDOMIZE_MEMORY_PHYSICAL_PADDING
 
 config HOTPLUG_CPU
 	def_bool y
-	depends on SMP
+	depends on SMP && !IPIPE
 
 config BOOTPARAM_HOTPLUG_CPU0
 	bool "Set default setting of cpu0_hotpluggable"
@@ -2222,7 +2235,7 @@ config COMPAT_VDSO
 choice
 	prompt "vsyscall table for legacy applications"
 	depends on X86_64
-	default LEGACY_VSYSCALL_EMULATE
+	default LEGACY_VSYSCALL_NONE
 	help
 	  Legacy user code that does not know how to find the vDSO expects
 	  to be able to issue three syscalls by calling fixed addresses in
@@ -2236,10 +2249,11 @@ choice
 	  static binaries, you can say None without a performance penalty
 	  to improve security.
 
-	  If unsure, select "Emulate".
+	  If unsure, select "None".
 
 	config LEGACY_VSYSCALL_NATIVE
 		bool "Native"
+		depends on !IPIPE
 		help
 		  Actual executable code is located in the fixed vsyscall
 		  address mapping, implementing time() efficiently. Since
@@ -2249,6 +2263,7 @@ choice
 
 	config LEGACY_VSYSCALL_EMULATE
 		bool "Emulate"
+		depends on !IPIPE
 		help
 		  The kernel traps and emulates calls into the fixed
 		  vsyscall address mapping. This makes the mapping
@@ -2313,6 +2328,7 @@ config CMDLINE_OVERRIDE
 
 config MODIFY_LDT_SYSCALL
 	bool "Enable the LDT (local descriptor table)" if EXPERT
+	depends on !IPIPE
 	default y
 	---help---
 	  Linux can allow user programs to install a per-process x86
@@ -2806,7 +2822,7 @@ source "fs/Kconfig.binfmt"
 
 config IA32_EMULATION
 	bool "IA32 Emulation"
-	depends on X86_64
+	depends on X86_64 && !IPIPE
 	select ARCH_WANT_OLD_COMPAT_IPC
 	select BINFMT_ELF
 	select COMPAT_BINFMT_ELF
@@ -2824,7 +2840,7 @@ config IA32_AOUT
 
 config X86_X32
 	bool "x32 ABI for 64-bit mode"
-	depends on X86_64
+	depends on X86_64 && !IPIPE
 	---help---
 	  Include code to run binaries for the x32 native 32-bit ABI
 	  for 64-bit processors.  An x32 process gets access to the
diff --git a/arch/x86/Kconfig.debug b/arch/x86/Kconfig.debug
index 6293a8768a91..9c5a69bbfd09 100644
--- a/arch/x86/Kconfig.debug
+++ b/arch/x86/Kconfig.debug
@@ -77,7 +77,7 @@ config X86_PTDUMP_CORE
 
 config X86_PTDUMP
 	tristate "Export kernel pagetable layout to userspace via debugfs"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	select DEBUG_FS
 	select X86_PTDUMP_CORE
 	---help---
@@ -136,7 +136,7 @@ config DOUBLEFAULT
 
 config DEBUG_TLBFLUSH
 	bool "Set upper limit of TLB entries to flush one-by-one"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	---help---
 
 	X86-only for now.
@@ -155,7 +155,7 @@ config DEBUG_TLBFLUSH
 
 config IOMMU_DEBUG
 	bool "Enable IOMMU debugging"
-	depends on GART_IOMMU && DEBUG_KERNEL
+	depends on DEBUG_KERNEL && GART_IOMMU && IOMMU
 	depends on X86_64
 	---help---
 	  Force the IOMMU to on even when you have less than 4GB of
@@ -171,6 +171,7 @@ config IOMMU_DEBUG
 
 config IOMMU_STRESS
 	bool "Enable IOMMU stress-test mode"
+	depends on IOMMU_DEBUG
 	---help---
 	  This option disables various optimizations in IOMMU related
 	  code to do real stress testing of the IOMMU code. This option
@@ -189,7 +190,7 @@ config HAVE_MMIOTRACE_SUPPORT
 
 config X86_DECODER_SELFTEST
 	bool "x86 instruction decoder selftest"
-	depends on DEBUG_KERNEL && KPROBES
+	depends on DEBUG_KERNEL && KPROBES && !IPIPE
 	depends on !COMPILE_TEST
 	---help---
 	 Perform x86 instruction decoder selftests at build time.
@@ -219,10 +220,10 @@ config IO_DELAY_TYPE_NONE
 
 choice
 	prompt "IO delay type"
-	default IO_DELAY_0X80
+	default IO_DELAY_NONE
 
 config IO_DELAY_0X80
-	bool "port 0x80 based port-IO delay [recommended]"
+	bool "port 0x80 based port-IO delay"
 	---help---
 	  This is the traditional Linux IO delay used for in/out_p.
 	  It is the most tested hence safest selection here.
@@ -240,7 +241,7 @@ config IO_DELAY_UDELAY
 	  while not having any side-effect on the IO port space.
 
 config IO_DELAY_NONE
-	bool "no port-IO delay"
+	bool "no port-IO delay [recommended]"
 	---help---
 	  No port-IO delay. Will break on old boxes that require port-IO
 	  delay for certain operations. Should work on most new machines.
@@ -273,14 +274,13 @@ endif
 
 config DEBUG_BOOT_PARAMS
 	bool "Debug boot parameters"
-	depends on DEBUG_KERNEL
-	depends on DEBUG_FS
+	depends on DEBUG_KERNEL && DEBUG_FS && !IPIPE
 	---help---
 	  This option will cause struct boot_params to be exported via debugfs.
 
 config CPA_DEBUG
 	bool "CPA self-test code"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	---help---
 	  Do change_page_attr() self-tests every 30 seconds.
 
@@ -300,7 +300,7 @@ config OPTIMIZE_INLINING
 
 config DEBUG_ENTRY
 	bool "Debug low-level entry code"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	---help---
 	  This option enables sanity checks in x86's low-level entry code.
 	  Some of these sanity checks may slow down kernel entries and
@@ -335,7 +335,7 @@ config DEBUG_IMR_SELFTEST
 
 config X86_DEBUG_FPU
 	bool "Debug the x86 FPU code"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	default y
 	---help---
 	  If this option is enabled then there will be extra sanity
@@ -347,7 +347,7 @@ config X86_DEBUG_FPU
 
 config PUNIT_ATOM_DEBUG
 	tristate "ATOM Punit debug driver"
-	depends on PCI
+	depends on PCI && !IPIPE
 	select DEBUG_FS
 	select IOSF_MBI
 	---help---
@@ -359,8 +359,7 @@ config PUNIT_ATOM_DEBUG
 
 choice
 	prompt "Choose kernel unwinder"
-	default UNWINDER_ORC if X86_64
-	default UNWINDER_FRAME_POINTER if X86_32
+	default UNWINDER_GUESS
 	---help---
 	  This determines which method will be used for unwinding kernel stack
 	  traces for panics, oopses, bugs, warnings, perf, /proc/<pid>/stack,
@@ -368,7 +367,7 @@ choice
 
 config UNWINDER_ORC
 	bool "ORC unwinder"
-	depends on X86_64
+	depends on X86_64 && !IPIPE
 	select STACK_VALIDATION
 	---help---
 	  This option enables the ORC (Oops Rewind Capability) unwinder for
@@ -384,6 +383,7 @@ config UNWINDER_ORC
 
 config UNWINDER_FRAME_POINTER
 	bool "Frame pointer unwinder"
+	depends on !IPIPE
 	select FRAME_POINTER
 	---help---
 	  This option enables the frame pointer unwinder for unwinding kernel
diff --git a/arch/x86/kvm/Kconfig b/arch/x86/kvm/Kconfig
index 3df51c287844..4971a5b221ac 100644
--- a/arch/x86/kvm/Kconfig
+++ b/arch/x86/kvm/Kconfig
@@ -8,6 +8,7 @@ source "virt/kvm/Kconfig"
 menuconfig VIRTUALIZATION
 	bool "Virtualization"
 	depends on HAVE_KVM || X86
+	depends on !IPIPE
 	default y
 	---help---
 	  Say Y here to get to see options for using your Linux host to run other
diff --git a/drivers/acpi/Kconfig b/drivers/acpi/Kconfig
index 5b1938f4b626..50eef4b2764f 100644
--- a/drivers/acpi/Kconfig
+++ b/drivers/acpi/Kconfig
@@ -60,6 +60,7 @@ config ACPI_CCA_REQUIRED
 
 config ACPI_DEBUGGER
 	bool "AML debugger interface"
+	depends on !IPIPE
 	select ACPI_DEBUG
 	help
 	  Enable in-kernel debugging of AML facilities: statistics,
@@ -126,6 +127,7 @@ config ACPI_REV_OVERRIDE_POSSIBLE
 
 config ACPI_EC_DEBUGFS
 	tristate "EC read/write access through /sys/kernel/debug/ec"
+	depends on !IPIPE
 	default n
 	help
 	  Say N to disable Embedded Controller /sys/kernel/debug interface
@@ -144,7 +146,7 @@ config ACPI_EC_DEBUGFS
 
 config ACPI_AC
 	tristate "AC Adapter"
-	depends on X86
+	depends on X86 && !IPIPE
 	select POWER_SUPPLY
 	default y
 	help
@@ -157,7 +159,7 @@ config ACPI_AC
 
 config ACPI_BATTERY
 	tristate "Battery"
-	depends on X86
+	depends on X86 && !IPIPE
 	select POWER_SUPPLY
 	default y
 	help
@@ -170,7 +172,7 @@ config ACPI_BATTERY
 
 config ACPI_BUTTON
 	tristate "Button"
-	depends on INPUT
+	depends on INPUT && !IPIPE
 	default y
 	help
 	  This driver handles events on the power, sleep, and lid buttons.
@@ -184,7 +186,7 @@ config ACPI_BUTTON
 config ACPI_VIDEO
 	tristate "Video"
 	depends on X86 && BACKLIGHT_CLASS_DEVICE
-	depends on INPUT
+	depends on INPUT && !IPIPE
 	select THERMAL
 	help
 	  This driver implements the ACPI Extensions For Display Adapters
@@ -209,6 +211,7 @@ config ACPI_FAN
 
 config ACPI_DOCK
 	bool "Dock"
+	depends on !IPIPE
 	help
 	  This driver supports ACPI-controlled docking stations and removable
 	  drive bays such as the IBM Ultrabay and the Dell Module Bay.
@@ -243,6 +246,7 @@ config ACPI_CPPC_LIB
 
 config ACPI_PROCESSOR
 	tristate "Processor"
+	depends on !IPIPE
 	depends on X86 || IA64 || ARM64
 	select ACPI_PROCESSOR_IDLE
 	select ACPI_CPU_FREQ_PSS if X86 || IA64
@@ -301,7 +305,7 @@ config ACPI_THERMAL
 config ACPI_NUMA
 	bool "NUMA support"
 	depends on NUMA
-	depends on (X86 || IA64 || ARM64)
+	depends on (X86 || IA64 || ARM64) && !IPIPE
 	default y if IA64_GENERIC || IA64_SGI_SN2 || ARM64
 
 config ACPI_CUSTOM_DSDT_FILE
@@ -336,6 +340,7 @@ config ACPI_TABLE_UPGRADE
 
 config ACPI_DEBUG
 	bool "Debug Statements"
+	depends on !IPIPE
 	default n
 	help
 	  The ACPI subsystem can produce debug output.  Saying Y enables this
@@ -348,7 +353,7 @@ config ACPI_DEBUG
 
 config ACPI_PCI_SLOT
 	bool "PCI slot detection driver"
-	depends on SYSFS
+	depends on SYSFS && !IPIPE
 	default n
 	help
 	  This driver creates entries in /sys/bus/pci/slots/ for all PCI
@@ -374,7 +379,7 @@ config X86_PM_TIMER
 
 config ACPI_CONTAINER
 	bool "Container and Module Devices"
-	default (ACPI_HOTPLUG_MEMORY || ACPI_HOTPLUG_CPU)
+	depends on (ACPI_HOTPLUG_MEMORY || ACPI_HOTPLUG_CPU)
 	help
 	  This driver supports ACPI Container and Module devices (IDs
 	  ACPI0004, PNP0A05, and PNP0A06).
@@ -408,7 +413,7 @@ config ACPI_HOTPLUG_IOAPIC
 
 config ACPI_SBS
 	tristate "Smart Battery System"
-	depends on X86
+	depends on X86 && !IPIPE
 	select POWER_SUPPLY
 	help
 	  This driver supports the Smart Battery System, another
@@ -419,6 +424,7 @@ config ACPI_SBS
 
 config ACPI_HED
 	tristate "Hardware Error Device"
+	depends on !IPIPE
 	help
 	  This driver supports the Hardware Error Device (PNP0C33),
 	  which is used to report some hardware errors notified via
@@ -441,7 +447,7 @@ config ACPI_CUSTOM_METHOD
 
 config ACPI_BGRT
 	bool "Boottime Graphics Resource Table support"
-	depends on EFI && (X86 || ARM64)
+	depends on EFI && (X86 || ARM64) && !IPIPE
         help
 	  This driver adds support for exposing the ACPI Boottime Graphics
 	  Resource Table, which allows the operating system to obtain
@@ -470,7 +476,7 @@ config ACPI_WATCHDOG
 
 config ACPI_EXTLOG
 	tristate "Extended Error Log support"
-	depends on X86_MCE && X86_LOCAL_APIC && EDAC
+	depends on X86_MCE && X86_LOCAL_APIC && EDAC && !IPIPE
 	select UEFI_CPER
 	default n
 	help
@@ -491,6 +497,7 @@ config ACPI_EXTLOG
 
 menuconfig PMIC_OPREGION
 	bool "PMIC (Power Management Integrated Circuit) operation region support"
+	depends on !IPIPE
 	help
 	  Select this option to enable support for ACPI operation
 	  region of the PMIC chip. The operation region can be used
@@ -526,6 +533,7 @@ endif
 
 config ACPI_CONFIGFS
 	tristate "ACPI configfs support"
+	depends on !IPIPE
 	select CONFIGFS_FS
 	help
 	  Select this option to enable support for ACPI configuration from
diff --git a/drivers/acpi/apei/Kconfig b/drivers/acpi/apei/Kconfig
index 52ae5438edeb..a871a3f546ae 100644
--- a/drivers/acpi/apei/Kconfig
+++ b/drivers/acpi/apei/Kconfig
@@ -10,7 +10,7 @@ config ACPI_APEI
 	select MISC_FILESYSTEMS
 	select PSTORE
 	select UEFI_CPER
-	depends on HAVE_ACPI_APEI
+	depends on HAVE_ACPI_APEI && !IPIPE
 	help
 	  APEI allows to report errors (for example from the chipset)
 	  to the operating system. This improves NMI handling
diff --git a/drivers/acpi/dptf/Kconfig b/drivers/acpi/dptf/Kconfig
index 90a2fd979282..bc4189889638 100644
--- a/drivers/acpi/dptf/Kconfig
+++ b/drivers/acpi/dptf/Kconfig
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 config DPTF_POWER
 	tristate "DPTF Platform Power Participant"
-	depends on X86
+	depends on X86 && !IPIPE
 	help
 	  This driver adds support for Dynamic Platform and Thermal Framework
 	  (DPTF) Platform Power Participant device (INT3407) support.
diff --git a/drivers/acpi/nfit/Kconfig b/drivers/acpi/nfit/Kconfig
index f7c57e33499e..1c32def97b0d 100644
--- a/drivers/acpi/nfit/Kconfig
+++ b/drivers/acpi/nfit/Kconfig
@@ -4,6 +4,7 @@ config ACPI_NFIT
 	depends on PHYS_ADDR_T_64BIT
 	depends on BLK_DEV
 	depends on ARCH_HAS_PMEM_API
+	depends on !IPIPE
 	select LIBNVDIMM
 	help
 	  Infrastructure to probe ACPI 6 compliant platforms for
diff --git a/drivers/char/ipmi/Kconfig b/drivers/char/ipmi/Kconfig
index f6fa056a52fc..764e68bf93b8 100644
--- a/drivers/char/ipmi/Kconfig
+++ b/drivers/char/ipmi/Kconfig
@@ -4,7 +4,7 @@
 
 menuconfig IPMI_HANDLER
        tristate 'IPMI top-level message handler'
-       depends on HAS_IOMEM
+       depends on HAS_IOMEM && !IPIPE
        select IPMI_DMI_DECODE if DMI
        help
          This enables the central IPMI message handler, required for IPMI
diff --git a/drivers/char/tpm/Kconfig b/drivers/char/tpm/Kconfig
index a30352202f1f..5a43a6a2877b 100644
--- a/drivers/char/tpm/Kconfig
+++ b/drivers/char/tpm/Kconfig
@@ -4,7 +4,7 @@
 
 menuconfig TCG_TPM
 	tristate "TPM Hardware Support"
-	depends on HAS_IOMEM
+	depends on HAS_IOMEM && !IPIPE
 	select SECURITYFS
 	select CRYPTO
 	select CRYPTO_HASH_INFO
diff --git a/drivers/cpufreq/Kconfig b/drivers/cpufreq/Kconfig
index d8addbce40bc..2ddefb18c5e1 100644
--- a/drivers/cpufreq/Kconfig
+++ b/drivers/cpufreq/Kconfig
@@ -2,6 +2,7 @@ menu "CPU Frequency scaling"
 
 config CPU_FREQ
 	bool "CPU Frequency scaling"
+	depends on !IPIPE
 	select SRCU
 	help
 	  CPU Frequency scaling allows you to change the clock speed of 
diff --git a/drivers/cpuidle/Kconfig b/drivers/cpuidle/Kconfig
index 7e48eb5bf0a7..5fc81c3570fc 100644
--- a/drivers/cpuidle/Kconfig
+++ b/drivers/cpuidle/Kconfig
@@ -2,6 +2,7 @@ menu "CPU Idle"
 
 config CPU_IDLE
 	bool "CPU idle PM support"
+	depends on !IPIPE
 	default y if ACPI || PPC_PSERIES
 	select CPU_IDLE_GOV_LADDER if (!NO_HZ && !NO_HZ_IDLE)
 	select CPU_IDLE_GOV_MENU if (NO_HZ || NO_HZ_IDLE)
diff --git a/drivers/crypto/Kconfig b/drivers/crypto/Kconfig
index 342bc777841c..0b22296d716e 100644
--- a/drivers/crypto/Kconfig
+++ b/drivers/crypto/Kconfig
@@ -1,6 +1,7 @@
 
 menuconfig CRYPTO_HW
 	bool "Hardware crypto devices"
+	depends on !IPIPE
 	default y
 	---help---
 	  Say Y here to get to see options for hardware crypto devices and
diff --git a/drivers/crypto/qat/Kconfig b/drivers/crypto/qat/Kconfig
index ce3cae40f949..a364b4cdeb86 100644
--- a/drivers/crypto/qat/Kconfig
+++ b/drivers/crypto/qat/Kconfig
@@ -47,7 +47,7 @@ config CRYPTO_DEV_QAT_C62X
 
 config CRYPTO_DEV_QAT_DH895xCCVF
 	tristate "Support for Intel(R) DH895xCC Virtual Function"
-	depends on X86 && PCI
+	depends on X86 && PCI && !IPIPE
 	select PCI_IOV
 	select CRYPTO_DEV_QAT
 
@@ -60,7 +60,7 @@ config CRYPTO_DEV_QAT_DH895xCCVF
 
 config CRYPTO_DEV_QAT_C3XXXVF
 	tristate "Support for Intel(R) C3XXX Virtual Function"
-	depends on X86 && PCI
+	depends on X86 && PCI && !IPIPE
 	select PCI_IOV
 	select CRYPTO_DEV_QAT
 	help
@@ -72,7 +72,7 @@ config CRYPTO_DEV_QAT_C3XXXVF
 
 config CRYPTO_DEV_QAT_C62XVF
 	tristate "Support for Intel(R) C62X Virtual Function"
-	depends on X86 && PCI
+	depends on X86 && PCI && !IPIPE
 	select PCI_IOV
 	select CRYPTO_DEV_QAT
 	help
diff --git a/drivers/devfreq/Kconfig b/drivers/devfreq/Kconfig
index 6a172d338f6d..a6a9466dd1f8 100644
--- a/drivers/devfreq/Kconfig
+++ b/drivers/devfreq/Kconfig
@@ -1,5 +1,6 @@
 menuconfig PM_DEVFREQ
 	bool "Generic Dynamic Voltage and Frequency Scaling (DVFS) support"
+	depends on !IPIPE
 	select SRCU
 	select PM_OPP
 	help
diff --git a/drivers/firmware/Kconfig b/drivers/firmware/Kconfig
index 6e4ed5a9c6fd..149640820c44 100644
--- a/drivers/firmware/Kconfig
+++ b/drivers/firmware/Kconfig
@@ -185,7 +185,7 @@ config RASPBERRYPI_FIRMWARE
 config FW_CFG_SYSFS
 	tristate "QEMU fw_cfg device support in sysfs"
 	depends on SYSFS && (ARM || ARM64 || PPC_PMAC || SPARC || X86)
-	depends on HAS_IOPORT_MAP
+	depends on HAS_IOPORT_MAP && !IPIPE
 	default n
 	help
 	  Say Y or M here to enable the exporting of the QEMU firmware
diff --git a/drivers/fsi/Kconfig b/drivers/fsi/Kconfig
index 6821ed0cd5e8..77fc81e47875 100644
--- a/drivers/fsi/Kconfig
+++ b/drivers/fsi/Kconfig
@@ -6,6 +6,7 @@ menu "FSI support"
 
 config FSI
 	tristate "FSI support"
+	depends on !IPIPE
 	select CRC4
 	---help---
 	  FSI - the FRU Support Interface - is a simple bus for low-level
diff --git a/drivers/gpu/drm/Kconfig b/drivers/gpu/drm/Kconfig
index 83cb2a88c204..0a0549e88ce0 100644
--- a/drivers/gpu/drm/Kconfig
+++ b/drivers/gpu/drm/Kconfig
@@ -282,7 +282,7 @@ source "drivers/gpu/drm/pl111/Kconfig"
 
 menuconfig DRM_LEGACY
 	bool "Enable legacy drivers (DANGEROUS)"
-	depends on DRM && MMU
+	depends on DRM && MMU && !IPIPE
 	select DRM_VM
 	help
 	  Enable legacy DRI1 drivers. Those drivers expose unsafe and dangerous
diff --git a/drivers/gpu/drm/gma500/Kconfig b/drivers/gpu/drm/gma500/Kconfig
index df11582f1efc..a611cfad3528 100644
--- a/drivers/gpu/drm/gma500/Kconfig
+++ b/drivers/gpu/drm/gma500/Kconfig
@@ -1,6 +1,6 @@
 config DRM_GMA500
 	tristate "Intel GMA5/600 KMS Framebuffer"
-	depends on DRM && PCI && X86 && MMU
+	depends on DRM && PCI && X86 && MMU && !IPIPE
 	select DRM_KMS_HELPER
 	select DRM_TTM
 	# GMA500 depends on ACPI_VIDEO when ACPI is enabled, just like i915
diff --git a/drivers/gpu/drm/i915/Kconfig b/drivers/gpu/drm/i915/Kconfig
index e9e64e8e9765..7cd3ff2ef066 100644
--- a/drivers/gpu/drm/i915/Kconfig
+++ b/drivers/gpu/drm/i915/Kconfig
@@ -1,7 +1,7 @@
 config DRM_I915
 	tristate "Intel 8xx/9xx/G3x/G4x/HD Graphics"
 	depends on DRM
-	depends on X86 && PCI
+	depends on X86 && PCI && !IPIPE
 	select INTEL_GTT
 	select INTERVAL_TREE
 	# we need shmfs for the swappable backing store, and in particular
diff --git a/drivers/gpu/drm/nouveau/Kconfig b/drivers/gpu/drm/nouveau/Kconfig
index a0c0630a0857..edf91194d887 100644
--- a/drivers/gpu/drm/nouveau/Kconfig
+++ b/drivers/gpu/drm/nouveau/Kconfig
@@ -1,6 +1,6 @@
 config DRM_NOUVEAU
 	tristate "Nouveau (NVIDIA) cards"
-	depends on DRM && PCI && MMU
+	depends on DRM && PCI && MMU && !IPIPE
         select FW_LOADER
 	select DRM_KMS_HELPER
 	select DRM_TTM
diff --git a/drivers/gpu/drm/qxl/Kconfig b/drivers/gpu/drm/qxl/Kconfig
index 378da5918e6c..65f5d8d45694 100644
--- a/drivers/gpu/drm/qxl/Kconfig
+++ b/drivers/gpu/drm/qxl/Kconfig
@@ -1,6 +1,6 @@
 config DRM_QXL
 	tristate "QXL virtual GPU"
-	depends on DRM && PCI && MMU
+	depends on DRM && PCI && MMU && !IPIPE
 	select DRM_KMS_HELPER
 	select DRM_TTM
 	select CRC32
diff --git a/drivers/gpu/drm/vmwgfx/Kconfig b/drivers/gpu/drm/vmwgfx/Kconfig
index 8c308dac99c5..c887be0cc07d 100644
--- a/drivers/gpu/drm/vmwgfx/Kconfig
+++ b/drivers/gpu/drm/vmwgfx/Kconfig
@@ -1,6 +1,6 @@
 config DRM_VMWGFX
 	tristate "DRM driver for VMware Virtual GPU"
-	depends on DRM && PCI && X86 && MMU
+	depends on DRM && PCI && X86 && MMU && !IPIPE
 	select FB_DEFERRED_IO
 	select FB_CFB_FILLRECT
 	select FB_CFB_COPYAREA
diff --git a/drivers/hwspinlock/Kconfig b/drivers/hwspinlock/Kconfig
index 6e0a5539a9ea..91fd47ae4c29 100644
--- a/drivers/hwspinlock/Kconfig
+++ b/drivers/hwspinlock/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig HWSPINLOCK
 	tristate "Hardware Spinlock drivers"
+	depends on !IPIPE
 
 config HWSPINLOCK_OMAP
 	tristate "OMAP Hardware Spinlock device"
diff --git a/drivers/infiniband/hw/usnic/Kconfig b/drivers/infiniband/hw/usnic/Kconfig
index 29ab11c34f3f..b59ea9838e9c 100644
--- a/drivers/infiniband/hw/usnic/Kconfig
+++ b/drivers/infiniband/hw/usnic/Kconfig
@@ -1,6 +1,7 @@
 config INFINIBAND_USNIC
 	tristate "Verbs support for Cisco VIC"
 	depends on NETDEVICES && ETHERNET && INET && PCI && INTEL_IOMMU
+	depends on !IPIPE
 	select ENIC
 	select NET_VENDOR_CISCO
 	select PCI_IOV
diff --git a/drivers/iommu/Kconfig b/drivers/iommu/Kconfig
index f3a21343e636..1e1ec7c3f97f 100644
--- a/drivers/iommu/Kconfig
+++ b/drivers/iommu/Kconfig
@@ -4,7 +4,7 @@ config IOMMU_API
 
 menuconfig IOMMU_SUPPORT
 	bool "IOMMU Hardware Support"
-	depends on MMU
+	depends on MMU && !IPIPE
 	default y
 	---help---
 	  Say Y here if you want to compile device drivers for IO Memory
diff --git a/drivers/mailbox/Kconfig b/drivers/mailbox/Kconfig
index c5731e5e3c6c..856e16ab82a9 100644
--- a/drivers/mailbox/Kconfig
+++ b/drivers/mailbox/Kconfig
@@ -1,5 +1,6 @@
 menuconfig MAILBOX
 	bool "Mailbox Hardware Support"
+	depends on !IPIPE
 	help
 	  Mailbox is a framework to control hardware communication between
 	  on-chip processors through queued messages and interrupt driven
diff --git a/drivers/md/Kconfig b/drivers/md/Kconfig
index 4a249ee86364..e93c84d486cc 100644
--- a/drivers/md/Kconfig
+++ b/drivers/md/Kconfig
@@ -251,7 +251,7 @@ config DM_DEBUG_BLOCK_MANAGER_LOCKING
 
 config DM_DEBUG_BLOCK_STACK_TRACING
        bool "Keep stack trace of persistent data block lock holders"
-       depends on STACKTRACE_SUPPORT && DM_DEBUG_BLOCK_MANAGER_LOCKING
+       depends on STACKTRACE_SUPPORT && DM_DEBUG_BLOCK_MANAGER_LOCKING && !IPIPE
        select STACKTRACE
        ---help---
 	 Enable this for messages that may help debug problems with the
diff --git a/drivers/misc/mic/Kconfig b/drivers/misc/mic/Kconfig
index 6fd9d367dea7..9b1f6f79e740 100644
--- a/drivers/misc/mic/Kconfig
+++ b/drivers/misc/mic/Kconfig
@@ -74,8 +74,7 @@ comment "Intel MIC Card Driver"
 config INTEL_MIC_CARD
 	tristate "Intel MIC Card Driver"
 	depends on 64BIT && X86
-	depends on INTEL_MIC_BUS && SCIF_BUS && MIC_COSM && VOP_BUS
-	select VIRTIO
+	depends on INTEL_MIC_BUS && SCIF_BUS && MIC_COSM && VOP_BUS && VIRTIO
 	help
 	  This enables card driver support for the Intel Many Integrated
 	  Core (MIC) device family. The card driver communicates shutdown/
@@ -130,9 +129,8 @@ comment "VOP Driver"
 
 config VOP
 	tristate "VOP Driver"
-	depends on 64BIT && PCI && X86 && VOP_BUS
+	depends on 64BIT && PCI && X86 && VOP_BUS && VIRTIO
 	select VHOST_RING
-	select VIRTIO
 	help
 	  This enables VOP (Virtio over PCIe) Driver support for the Intel
 	  Many Integrated Core (MIC) family of PCIe form factor coprocessor
diff --git a/drivers/misc/vmw_vmci/Kconfig b/drivers/misc/vmw_vmci/Kconfig
index 39c2ecadb273..34de99aff944 100644
--- a/drivers/misc/vmw_vmci/Kconfig
+++ b/drivers/misc/vmw_vmci/Kconfig
@@ -4,7 +4,7 @@
 
 config VMWARE_VMCI
 	tristate "VMware VMCI Driver"
-	depends on X86 && PCI
+	depends on X86 && PCI && !IPIPE
 	help
 	  This is VMware's Virtual Machine Communication Interface.  It enables
 	  high-speed communication between host and guest in a virtual
diff --git a/drivers/net/caif/Kconfig b/drivers/net/caif/Kconfig
index f81df91a9ce1..7b38ada49c0c 100644
--- a/drivers/net/caif/Kconfig
+++ b/drivers/net/caif/Kconfig
@@ -43,9 +43,8 @@ config CAIF_HSI
 
 config CAIF_VIRTIO
 	tristate "CAIF virtio transport driver"
-	depends on CAIF && HAS_DMA
+	depends on CAIF && HAS_DMA && VIRTIO
 	select VHOST_RING
-	select VIRTIO
 	select GENERIC_ALLOCATOR
 	default n
 	---help---
diff --git a/drivers/pci/Kconfig b/drivers/pci/Kconfig
index c32a77fc8b03..22d8ce2c0b45 100644
--- a/drivers/pci/Kconfig
+++ b/drivers/pci/Kconfig
@@ -41,7 +41,7 @@ config PCI_DEBUG
 
 config PCI_REALLOC_ENABLE_AUTO
 	bool "Enable PCI resource re-allocation detection"
-	depends on PCI
+	depends on PCI && !IPIPE
 	help
 	  Say Y here if you want the PCI core to detect if PCI resource
 	  re-allocation needs to be enabled. You can always use pci=realloc=on
@@ -91,7 +91,7 @@ config PCI_LOCKLESS_CONFIG
 
 config PCI_IOV
 	bool "PCI IOV support"
-	depends on PCI
+	depends on PCI && !IPIPE
 	select PCI_ATS
 	help
 	  I/O Virtualization is a PCI feature supported by some devices
@@ -102,7 +102,7 @@ config PCI_IOV
 
 config PCI_PRI
 	bool "PCI PRI support"
-	depends on PCI
+	depends on PCI && !IPIPE
 	select PCI_ATS
 	help
 	  PRI is the PCI Page Request Interface. It allows PCI devices that are
@@ -112,7 +112,7 @@ config PCI_PRI
 
 config PCI_PASID
 	bool "PCI PASID support"
-	depends on PCI
+	depends on PCI && !IPIPE
 	select PCI_ATS
 	help
 	  Process Address Space Identifiers (PASIDs) can be used by PCI devices
@@ -130,6 +130,7 @@ config PCI_LABEL
 config PCI_HYPERV
         tristate "Hyper-V PCI Frontend"
         depends on PCI && X86 && HYPERV && PCI_MSI && PCI_MSI_IRQ_DOMAIN && X86_64
+	depends on !IPIPE
         help
           The PCI device frontend driver allows the kernel to import arbitrary
           PCI devices from a PCI backend to support PCI driver domains.
diff --git a/drivers/pci/hotplug/Kconfig b/drivers/pci/hotplug/Kconfig
index aadce45a9b4a..6d6caf6f0c38 100644
--- a/drivers/pci/hotplug/Kconfig
+++ b/drivers/pci/hotplug/Kconfig
@@ -4,7 +4,7 @@
 
 menuconfig HOTPLUG_PCI
 	bool "Support for PCI Hotplug"
-	depends on PCI && SYSFS
+	depends on PCI && SYSFS && !IPIPE
 	---help---
 	  Say Y here if you have a motherboard with a PCI Hotplug controller.
 	  This allows you to add and remove PCI cards while the machine is
diff --git a/drivers/pci/pcie/Kconfig b/drivers/pci/pcie/Kconfig
index ac53edbc9613..8268928edb4b 100644
--- a/drivers/pci/pcie/Kconfig
+++ b/drivers/pci/pcie/Kconfig
@@ -3,7 +3,7 @@
 #
 config PCIEPORTBUS
 	bool "PCI Express Port Bus support"
-	depends on PCI
+	depends on PCI && !IPIPE
 	help
 	  This automatically enables PCI Express Port Bus support. Users can
 	  choose Native Hot-Plug support, Advanced Error Reporting support,
diff --git a/drivers/platform/x86/Kconfig b/drivers/platform/x86/Kconfig
index 09035705d0a0..a7b3cbe24329 100644
--- a/drivers/platform/x86/Kconfig
+++ b/drivers/platform/x86/Kconfig
@@ -627,7 +627,7 @@ config ASUS_WIRELESS
 
 config ACPI_WMI
 	tristate "WMI"
-	depends on ACPI
+	depends on ACPI && !IPIPE
 	help
 	  This driver adds support for the ACPI-WMI (Windows Management
 	  Instrumentation) mapper device (PNP0C14) found on some systems.
diff --git a/drivers/power/avs/Kconfig b/drivers/power/avs/Kconfig
index a67eeace6a89..5d2d8bc3653d 100644
--- a/drivers/power/avs/Kconfig
+++ b/drivers/power/avs/Kconfig
@@ -1,5 +1,6 @@
 menuconfig POWER_AVS
 	bool "Adaptive Voltage Scaling class support"
+	depends on !IPIPE
 	help
 	  AVS is a power management technique which finely controls the
 	  operating voltage of a device in order to optimize (i.e. reduce)
diff --git a/drivers/power/reset/Kconfig b/drivers/power/reset/Kconfig
index ca0de1a78e85..6b1312d9d1f9 100644
--- a/drivers/power/reset/Kconfig
+++ b/drivers/power/reset/Kconfig
@@ -1,5 +1,6 @@
 menuconfig POWER_RESET
 	bool "Board level reset or power off"
+	depends on !IPIPE
 	help
 	  Provides a number of drivers which either reset a complete board
 	  or shut it down, by manipulating the main power supply on the board.
diff --git a/drivers/powercap/Kconfig b/drivers/powercap/Kconfig
index 85727ef6ce8e..196119270fde 100644
--- a/drivers/powercap/Kconfig
+++ b/drivers/powercap/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig POWERCAP
 	bool "Generic powercap sysfs driver"
+	depends on !IPIPE
 	help
 	  The power capping sysfs interface allows kernel subsystems to expose power
 	  capping settings to user space in a consistent way.  Usually, it consists
diff --git a/drivers/pwm/Kconfig b/drivers/pwm/Kconfig
index 763ee50ea57d..11a1fb3e8def 100644
--- a/drivers/pwm/Kconfig
+++ b/drivers/pwm/Kconfig
@@ -1,5 +1,6 @@
 menuconfig PWM
 	bool "Pulse-Width Modulation (PWM) Support"
+	depends on !IPIPE
 	help
 	  Generic Pulse-Width Modulation (PWM) support.
 
diff --git a/drivers/regulator/Kconfig b/drivers/regulator/Kconfig
index 0fd6195601ba..0d08d64b1718 100644
--- a/drivers/regulator/Kconfig
+++ b/drivers/regulator/Kconfig
@@ -1,5 +1,6 @@
 menuconfig REGULATOR
 	bool "Voltage and Current Regulator Support"
+	depends on !IPIPE
 	help
 	  Generic Voltage and Current Regulator support.
 
diff --git a/drivers/remoteproc/Kconfig b/drivers/remoteproc/Kconfig
index bf04479456a0..50c261ae8c89 100644
--- a/drivers/remoteproc/Kconfig
+++ b/drivers/remoteproc/Kconfig
@@ -2,10 +2,9 @@ menu "Remoteproc drivers"
 
 config REMOTEPROC
 	tristate "Support for Remote Processor subsystem"
-	depends on HAS_DMA
+	depends on HAS_DMA && VIRTIO
 	select CRC32
 	select FW_LOADER
-	select VIRTIO
 	help
 	  Support for remote processors (such as DSP coprocessors). These
 	  are mainly used on embedded systems.
diff --git a/drivers/reset/Kconfig b/drivers/reset/Kconfig
index e2baecbb9dd3..8f5c51c2e30a 100644
--- a/drivers/reset/Kconfig
+++ b/drivers/reset/Kconfig
@@ -3,7 +3,7 @@ config ARCH_HAS_RESET_CONTROLLER
 
 menuconfig RESET_CONTROLLER
 	bool "Reset Controller Support"
-	default y if ARCH_HAS_RESET_CONTROLLER
+	default y if ARCH_HAS_RESET_CONTROLLER && !IPIPE
 	help
 	  Generic Reset Controller support.
 
diff --git a/drivers/rpmsg/Kconfig b/drivers/rpmsg/Kconfig
index 0fe6eac46512..2a54818008bc 100644
--- a/drivers/rpmsg/Kconfig
+++ b/drivers/rpmsg/Kconfig
@@ -48,7 +48,7 @@ config RPMSG_QCOM_SMD
 
 config RPMSG_VIRTIO
 	tristate
+	depends on VIRTIO
 	select RPMSG
-	select VIRTIO
 
 endmenu
diff --git a/drivers/scsi/ufs/Kconfig b/drivers/scsi/ufs/Kconfig
index e27b4d4e6ae2..1b6017170e58 100644
--- a/drivers/scsi/ufs/Kconfig
+++ b/drivers/scsi/ufs/Kconfig
@@ -34,7 +34,7 @@
 
 config SCSI_UFSHCD
 	tristate "Universal Flash Storage Controller Driver Core"
-	depends on SCSI && SCSI_DMA
+	depends on SCSI && SCSI_DMA && !IPIPE
 	select PM_DEVFREQ
 	select DEVFREQ_GOV_SIMPLE_ONDEMAND
 	select NLS
diff --git a/drivers/thermal/Kconfig b/drivers/thermal/Kconfig
index e3f0d1fd1720..6bca6fc8e643 100644
--- a/drivers/thermal/Kconfig
+++ b/drivers/thermal/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig THERMAL
 	tristate "Generic Thermal sysfs driver"
+	depends on !IPIPE
 	help
 	  Generic Thermal Sysfs driver offers a generic mechanism for
 	  thermal management. Usually it's made up of one or more thermal
diff --git a/drivers/virt/Kconfig b/drivers/virt/Kconfig
index 99ebdde590f8..61e7cdb69899 100644
--- a/drivers/virt/Kconfig
+++ b/drivers/virt/Kconfig
@@ -4,6 +4,7 @@
 
 menuconfig VIRT_DRIVERS
 	bool "Virtualization drivers"
+	depends on !IPIPE
 	---help---
 	  Say Y here to get to see options for device drivers that support
 	  virtualization environments.
diff --git a/drivers/virtio/Kconfig b/drivers/virtio/Kconfig
index cff773f15b7e..e5bfe34114fa 100644
--- a/drivers/virtio/Kconfig
+++ b/drivers/virtio/Kconfig
@@ -1,5 +1,6 @@
 config VIRTIO
 	tristate
+	depends on !IPIPE
 	---help---
 	  This option is selected by any driver which implements the virtio
 	  bus, such as CONFIG_VIRTIO_PCI, CONFIG_VIRTIO_MMIO, CONFIG_RPMSG
@@ -9,8 +10,7 @@ menu "Virtio drivers"
 
 config VIRTIO_PCI
 	tristate "PCI driver for virtio devices"
-	depends on PCI
-	select VIRTIO
+	depends on PCI && VIRTIO
 	---help---
 	  This driver provides support for virtio based paravirtual device
 	  drivers over PCI.  This requires that your VMM has appropriate PCI
@@ -60,8 +60,7 @@ config VIRTIO_INPUT
 
  config VIRTIO_MMIO
 	tristate "Platform bus driver for memory mapped virtio devices"
-	depends on HAS_IOMEM && HAS_DMA
- 	select VIRTIO
+	depends on HAS_IOMEM && HAS_DMA && VIRTIO
  	---help---
  	 This drivers provides support for memory mapped virtio
 	 platform device driver.
diff --git a/drivers/watchdog/Kconfig b/drivers/watchdog/Kconfig
index f55328a31629..2e3b670c3201 100644
--- a/drivers/watchdog/Kconfig
+++ b/drivers/watchdog/Kconfig
@@ -5,6 +5,7 @@
 
 menuconfig WATCHDOG
 	bool "Watchdog Timer Support"
+	depends on !IPIPE
 	---help---
 	  If you say Y here (and to one of the following options) and create a
 	  character special file /dev/watchdog with major number 10 and minor
diff --git a/fs/Kconfig.binfmt b/fs/Kconfig.binfmt
index b2f82cf6bf86..2debf2a84334 100644
--- a/fs/Kconfig.binfmt
+++ b/fs/Kconfig.binfmt
@@ -157,6 +157,7 @@ config BINFMT_EM86
 
 config BINFMT_MISC
 	tristate "Kernel support for MISC binaries"
+	depends on !IPIPE
 	---help---
 	  If you say Y here, it will be possible to plug wrapper-driven binary
 	  formats into the kernel. You will like this especially when you use
diff --git a/include/linux/ipipe.h b/include/linux/ipipe.h
index b8a80eeade4f..523ff646853a 100644
--- a/include/linux/ipipe.h
+++ b/include/linux/ipipe.h
@@ -30,7 +30,7 @@
 #include <linux/thread_info.h>
 #include <linux/ipipe_debug.h>
 #include <asm/ptrace.h>
-#ifdef CONFIG_HAVE_IPIPE_SUPPORT
+#ifdef CONFIG_IPIPE
 #include <asm/ipipe.h>
 #endif
 
diff --git a/init/Kconfig b/init/Kconfig
index 33cca939377e..a94cbc0bf838 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -298,6 +298,7 @@ config FHANDLE
 
 config USELIB
 	bool "uselib syscall"
+	depends on !IPIPE
 	def_bool ALPHA || M68K || SPARC || X86_32 || IA32_EMULATION
 	help
 	  This option enables the uselib syscall, a system call used in the
@@ -308,7 +309,7 @@ config USELIB
 
 config AUDIT
 	bool "Auditing support"
-	depends on NET
+	depends on NET && !IPIPE
 	help
 	  Enable auditing infrastructure that can be used with another
 	  kernel subsystem, such as SELinux (which requires this for
@@ -373,6 +374,7 @@ config VIRT_CPU_ACCOUNTING_GEN
 	bool "Full dynticks CPU time accounting"
 	depends on HAVE_CONTEXT_TRACKING
 	depends on HAVE_VIRT_CPU_ACCOUNTING_GEN
+	depends on !IPIPE
 	select VIRT_CPU_ACCOUNTING
 	select CONTEXT_TRACKING
 	help
@@ -392,6 +394,7 @@ endchoice
 config IRQ_TIME_ACCOUNTING
 	bool "Fine granularity task level IRQ time accounting"
 	depends on HAVE_IRQ_TIME_ACCOUNTING && !VIRT_CPU_ACCOUNTING_NATIVE
+	depends on !IPIPE
 	help
 	  Select this option to enable fine granularity task irq time
 	  accounting. This is done by reading a timestamp on each
@@ -718,6 +721,7 @@ config CGROUP_WRITEBACK
 
 menuconfig CGROUP_SCHED
 	bool "CPU controller"
+	depends on !IPIPE
 	default n
 	help
 	  This feature lets CPU scheduler recognize task groups and control CPU
@@ -809,7 +813,7 @@ config CGROUP_HUGETLB
 
 config CPUSETS
 	bool "Cpuset controller"
-	depends on SMP
+	depends on SMP && !IPIPE
 	help
 	  This option will let you create and manage CPUSETs which
 	  allow dynamically partitioning a system into sets of CPUs and
@@ -837,7 +841,7 @@ config CGROUP_CPUACCT
 
 config CGROUP_PERF
 	bool "Perf controller"
-	depends on PERF_EVENTS
+	depends on PERF_EVENTS && !IPIPE
 	help
 	  This option extends the perf per-cpu mode to restrict monitoring
 	  to threads which belong to the cgroup specified and run on the
@@ -861,7 +865,7 @@ config CGROUP_BPF
 config CGROUP_DEBUG
 	bool "Debug controller"
 	default n
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  This option enables a simple controller that exports
 	  debugging information about the cgroups framework. This
@@ -949,6 +953,7 @@ endif # NAMESPACES
 
 config SCHED_AUTOGROUP
 	bool "Automatic process group scheduling"
+	depends on !IPIPE
 	select CGROUPS
 	select CGROUP_SCHED
 	select FAIR_GROUP_SCHED
@@ -1044,6 +1049,7 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size"
+	depends on !IPIPE
 	help
 	  Enabling this option will pass "-Os" instead of "-O2" to
 	  your compiler resulting in a smaller kernel.
@@ -1174,16 +1180,16 @@ config POSIX_TIMERS
 	  If unsure say y.
 
 config KALLSYMS
-	 bool "Load all symbols for debugging/ksymoops" if EXPERT
-	 default y
-	 help
+	bool "Load all symbols for debugging/ksymoops" if EXPERT
+	default y
+	help
 	   Say Y here to let the kernel print out symbolic crash information and
 	   symbolic stack backtraces. This increases the size of the kernel
 	   somewhat, as all symbols have to be loaded into the kernel image.
 
 config KALLSYMS_ALL
 	bool "Include all symbols in kallsyms"
-	depends on DEBUG_KERNEL && KALLSYMS
+	depends on DEBUG_KERNEL && KALLSYMS && !IPIPE
 	help
 	   Normally kallsyms only contains the symbols of functions for nicer
 	   OOPS messages and backtraces (i.e., symbols from the text and inittext
@@ -1504,7 +1510,7 @@ config VM_EVENT_COUNTERS
 config SLUB_DEBUG
 	default y
 	bool "Enable SLUB debugging support" if EXPERT
-	depends on SLUB && SYSFS
+	depends on SLUB && SYSFS && !IPIPE
 	help
 	  SLUB has extensive debug support features. Disabling these can
 	  result in significant savings in code size. This also disables
@@ -1514,7 +1520,7 @@ config SLUB_DEBUG
 config SLUB_MEMCG_SYSFS_ON
 	default n
 	bool "Enable memcg SLUB sysfs support by default" if EXPERT
-	depends on SLUB && SYSFS && MEMCG
+	depends on SLUB && SYSFS && MEMCG && !IPIPE
 	help
 	  SLUB creates a directory under /sys/kernel/slab for each
 	  allocation cache to host info and debug files. If memory
@@ -1606,7 +1612,7 @@ config SLAB_FREELIST_HARDENED
 
 config SLUB_CPU_PARTIAL
 	default y
-	depends on SLUB && SMP
+	depends on SLUB && SMP && !IPIPE
 	bool "SLUB per cpu partial cache"
 	help
 	  Per cpu partial caches accellerate objects allocation and freeing
@@ -1741,6 +1747,7 @@ config MODULE_FORCE_UNLOAD
 
 config MODVERSIONS
 	bool "Module versioning support"
+	depends on !IPIPE
 	help
 	  Usually, you have to use modules compiled with your kernel.
 	  Saying Y here makes it sometimes possible to use modules
@@ -1766,7 +1773,7 @@ config MODULE_SRCVERSION_ALL
 
 config MODULE_SIG
 	bool "Module signature verification"
-	depends on MODULES
+	depends on MODULES && !IPIPE
 	select SYSTEM_DATA_VERIFICATION
 	help
 	  Check modules for valid signatures upon load: the signature
@@ -1881,7 +1888,7 @@ endchoice
 
 config TRIM_UNUSED_KSYMS
 	bool "Trim unused exported kernel symbols"
-	depends on MODULES && !UNUSED_SYMBOLS
+	depends on MODULES && !UNUSED_SYMBOLS && !IPIPE
 	help
 	  The kernel and some modules make many symbols available for
 	  other modules to use via EXPORT_SYMBOL() and variants. Depending
diff --git a/kernel/ipipe/Kconfig b/kernel/ipipe/Kconfig
index d17edb89f1f5..c77f2255e175 100644
--- a/kernel/ipipe/Kconfig
+++ b/kernel/ipipe/Kconfig
@@ -1,12 +1,17 @@
-
-config HAVE_IPIPE_SUPPORT
-       depends on GENERIC_CLOCKEVENTS
-       bool
-
 config IPIPE
 	bool "Interrupt pipeline"
-	depends on HAVE_IPIPE_SUPPORT
-	default n
+	depends on X86_64
+	default y
+	select BUG
+	select PRINTK
+	select KALLSYMS
+	select PANIC_ON_OOPS
+	select DEBUG_BUGVERBOSE
+	select PROC_FS
+	select SCHED_OMIT_FRAME_POINTER
+	select HIGH_RES_TIMERS
+	select X86_IO_APIC
+	select X86_UP_IOAPIC if !SMP
 	---help---
 	  Activate this option if you want the interrupt pipeline to be
 	  compiled in.
diff --git a/kernel/irq/Kconfig b/kernel/irq/Kconfig
index a117adf7084b..ea84542fcc6f 100644
--- a/kernel/irq/Kconfig
+++ b/kernel/irq/Kconfig
@@ -99,7 +99,7 @@ config IRQ_TIMINGS
 
 config IRQ_DOMAIN_DEBUG
 	bool "Expose hardware/virtual IRQ mapping via debugfs"
-	depends on IRQ_DOMAIN && DEBUG_FS
+	depends on IRQ_DOMAIN && DEBUG_FS && !IPIPE
 	help
 	  This option will show the mapping relationship between hardware irq
 	  numbers and Linux irq numbers. The mapping is exposed via debugfs
@@ -126,7 +126,7 @@ config SPARSE_IRQ
 
 config GENERIC_IRQ_DEBUGFS
 	bool "Expose irq internals in debugfs"
-	depends on DEBUG_FS
+	depends on DEBUG_FS && !IPIPE
 	default n
 	---help---
 
diff --git a/kernel/power/Kconfig b/kernel/power/Kconfig
index dd2b5a4d89a5..00007dd41191 100644
--- a/kernel/power/Kconfig
+++ b/kernel/power/Kconfig
@@ -1,6 +1,6 @@
 config SUSPEND
 	bool "Suspend to RAM and standby"
-	depends on ARCH_SUSPEND_POSSIBLE
+	depends on ARCH_SUSPEND_POSSIBLE && !IPIPE
 	default y
 	---help---
 	  Allow the system to enter sleep states in which main memory is
@@ -29,11 +29,12 @@ config SUSPEND_SKIP_SYNC
 	  user-space before invoking suspend.  Say Y if that's your case.
 
 config HIBERNATE_CALLBACKS
+	depends on !IPIPE
 	bool
 
 config HIBERNATION
 	bool "Hibernation (aka 'suspend to disk')"
-	depends on SWAP && ARCH_HIBERNATION_POSSIBLE
+	depends on SWAP && ARCH_HIBERNATION_POSSIBLE && !IPIPE
 	select HIBERNATE_CALLBACKS
 	select LZO_COMPRESS
 	select LZO_DECOMPRESS
@@ -143,6 +144,7 @@ config PM_WAKELOCKS_GC
 
 config PM
 	bool "Device power management core functionality"
+	depends on !IPIPE
 	---help---
 	  Enable functionality allowing I/O devices to be put into energy-saving
 	  (low power) states, for example after a specified period of inactivity
diff --git a/kernel/rcu/Kconfig.debug b/kernel/rcu/Kconfig.debug
index 2034d02e8ee3..bd33c2767fe2 100644
--- a/kernel/rcu/Kconfig.debug
+++ b/kernel/rcu/Kconfig.debug
@@ -13,7 +13,7 @@ config TORTURE_TEST
 
 config RCU_PERF_TEST
 	tristate "performance tests for RCU"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	select TORTURE_TEST
 	select SRCU
 	select TASKS_RCU
@@ -30,7 +30,7 @@ config RCU_PERF_TEST
 
 config RCU_TORTURE_TEST
 	tristate "torture tests for RCU"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	select TORTURE_TEST
 	select SRCU
 	select TASKS_RCU
@@ -58,7 +58,7 @@ config RCU_CPU_STALL_TIMEOUT
 
 config RCU_TRACE
 	bool "Enable tracing for RCU"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	default y if TREE_RCU
 	select TRACE_CLOCK
 	help
@@ -70,7 +70,7 @@ config RCU_TRACE
 
 config RCU_EQS_DEBUG
 	bool "Provide debugging asserts for adding NO_HZ support to an arch"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  This option provides consistency checks in RCU's handling of
 	  NO_HZ.  These checks have proven quite helpful in detecting
diff --git a/kernel/trace/Kconfig b/kernel/trace/Kconfig
index 5f93356dd495..5af0e1637403 100644
--- a/kernel/trace/Kconfig
+++ b/kernel/trace/Kconfig
@@ -122,7 +122,8 @@ if TRACING_SUPPORT
 
 menuconfig FTRACE
 	bool "Tracers"
-	default y if DEBUG_KERNEL
+	depends on !IPIPE || IPIPE_TRACE
+	default n if DEBUG_KERNEL
 	help
 	  Enable the kernel tracing infrastructure.
 
diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index d1057c94ae0d..5abf9f86501a 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -47,7 +47,7 @@ config MESSAGE_LOGLEVEL_DEFAULT
 
 config BOOT_PRINTK_DELAY
 	bool "Delay each boot printk message by N milliseconds"
-	depends on DEBUG_KERNEL && PRINTK && GENERIC_CALIBRATE_DELAY
+	depends on DEBUG_KERNEL && PRINTK && GENERIC_CALIBRATE_DELAY && !IPIPE
 	help
 	  This build option allows you to read kernel boot messages
 	  by inserting a short delay after each one.  The delay is
@@ -68,6 +68,7 @@ config DYNAMIC_DEBUG
 	default n
 	depends on PRINTK
 	depends on DEBUG_FS
+	depends on !IPIPE
 	help
 
 	  Compiles debug level messages into the kernel, which would not
@@ -139,7 +140,7 @@ menu "Compile-time checks and compiler options"
 
 config DEBUG_INFO
 	bool "Compile the kernel with debug info"
-	depends on DEBUG_KERNEL && !COMPILE_TEST
+	depends on DEBUG_KERNEL && !COMPILE_TEST && !IPIPE
 	help
           If you say Y here the resulting kernel image will include
 	  debugging info resulting in a larger kernel image.
@@ -230,7 +231,7 @@ config FRAME_WARN
 
 config STRIP_ASM_SYMS
 	bool "Strip assembler-generated symbols during link"
-	default n
+	default y
 	help
 	  Strip internal assembler-generated symbols during a link (symbols
 	  that look like '.Lxxx') so they don't pollute the output of
@@ -238,7 +239,7 @@ config STRIP_ASM_SYMS
 
 config READABLE_ASM
         bool "Generate readable assembler code"
-        depends on DEBUG_KERNEL
+        depends on DEBUG_KERNEL && !IPIPE
         help
           Disable some compiler optimizations that tend to generate human unreadable
           assembler output. This may make the kernel slightly slower, but it helps
@@ -247,7 +248,8 @@ config READABLE_ASM
 
 config UNUSED_SYMBOLS
 	bool "Enable unused/obsolete exported symbols"
-	default y if X86
+	depends on !IPIPE
+	default n if X86
 	help
 	  Unused but exported symbols make the kernel needlessly bigger.  For
 	  that reason most of these unused exports will soon be removed.  This
@@ -263,7 +265,7 @@ config UNUSED_SYMBOLS
 
 config PAGE_OWNER
 	bool "Track page owner"
-	depends on DEBUG_KERNEL && STACKTRACE_SUPPORT
+	depends on DEBUG_KERNEL && STACKTRACE_SUPPORT && !IPIPE
 	select DEBUG_FS
 	select STACKTRACE
 	select STACKDEPOT
@@ -307,6 +309,7 @@ config HEADERS_CHECK
 
 config DEBUG_SECTION_MISMATCH
 	bool "Enable full Section mismatch analysis"
+	depends on !IPIPE
 	help
 	  The section mismatch analysis checks if there are illegal
 	  references from one section to another section.
@@ -360,6 +363,7 @@ config FRAME_POINTER
 		(CRIS || M68K || FRV || UML || \
 		 SUPERH || BLACKFIN || MN10300 || METAG) || \
 		ARCH_WANT_FRAME_POINTERS
+	depends on !IPIPE
 	default y if (DEBUG_INFO && UML) || ARCH_WANT_FRAME_POINTERS
 	help
 	  If you say Y here the resulting kernel image will be slightly
@@ -368,7 +372,7 @@ config FRAME_POINTER
 
 config STACK_VALIDATION
 	bool "Compile-time stack metadata validation"
-	depends on HAVE_STACK_VALIDATION
+	depends on HAVE_STACK_VALIDATION && !IPIPE
 	default n
 	help
 	  Add compile-time checks to validate stack metadata, including frame
@@ -383,7 +387,7 @@ config STACK_VALIDATION
 
 config DEBUG_FORCE_WEAK_PER_CPU
 	bool "Force weak per-cpu definitions"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  s390 and alpha require percpu variables in modules to be
 	  defined weak to work around addressing range issue which
@@ -446,7 +450,7 @@ source mm/Kconfig.debug
 
 config DEBUG_OBJECTS
 	bool "Debug object operations"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  If you say Y here, additional code will be inserted into the
 	  kernel to track the life time of various objects and validate
@@ -507,7 +511,7 @@ config DEBUG_OBJECTS_ENABLE_DEFAULT
 
 config DEBUG_SLAB
 	bool "Debug slab memory allocations"
-	depends on DEBUG_KERNEL && SLAB
+	depends on DEBUG_KERNEL && SLAB && !IPIPE
 	help
 	  Say Y here to have the kernel do limited verification on memory
 	  allocation as well as poisoning memory on free to catch use of freed
@@ -533,7 +537,7 @@ config SLUB_DEBUG_ON
 config SLUB_STATS
 	default n
 	bool "Enable SLUB performance statistics"
-	depends on SLUB && SYSFS
+	depends on SLUB && SYSFS && !IPIPE
 	help
 	  SLUB statistics are useful to debug SLUBs allocation behavior in
 	  order find ways to optimize the allocator. This should never be
@@ -548,7 +552,7 @@ config HAVE_DEBUG_KMEMLEAK
 
 config DEBUG_KMEMLEAK
 	bool "Kernel memory leak detector"
-	depends on DEBUG_KERNEL && HAVE_DEBUG_KMEMLEAK
+	depends on DEBUG_KERNEL && HAVE_DEBUG_KMEMLEAK && !IPIPE
 	select DEBUG_FS
 	select STACKTRACE if STACKTRACE_SUPPORT
 	select KALLSYMS
@@ -598,7 +602,7 @@ config DEBUG_KMEMLEAK_DEFAULT_OFF
 
 config DEBUG_STACK_USAGE
 	bool "Stack utilization instrumentation"
-	depends on DEBUG_KERNEL && !IA64
+	depends on DEBUG_KERNEL && !IA64 && !IPIPE
 	help
 	  Enables the display of the minimum amount of free stack which each
 	  task has ever had available in the sysrq-T and sysrq-P debug output.
@@ -607,7 +611,7 @@ config DEBUG_STACK_USAGE
 
 config DEBUG_VM
 	bool "Debug VM"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  Enable this to turn on extended checks in the virtual-memory system
           that may impact performance.
@@ -645,7 +649,7 @@ config ARCH_HAS_DEBUG_VIRTUAL
 
 config DEBUG_VIRTUAL
 	bool "Debug VM translations"
-	depends on DEBUG_KERNEL && ARCH_HAS_DEBUG_VIRTUAL
+	depends on DEBUG_KERNEL && ARCH_HAS_DEBUG_VIRTUAL && !IPIPE
 	help
 	  Enable some costly sanity checks in virtual to page code. This can
 	  catch mistakes with virt_to_page() and friends.
@@ -698,6 +702,7 @@ config DEBUG_PER_CPU_MAPS
 	bool "Debug access to per_cpu maps"
 	depends on DEBUG_KERNEL
 	depends on SMP
+	depends on !IPIPE
 	help
 	  Say Y to verify that the per_cpu map being accessed has
 	  been set up. This adds a fair amount of code to kernel memory
@@ -707,7 +712,7 @@ config DEBUG_PER_CPU_MAPS
 
 config DEBUG_HIGHMEM
 	bool "Highmem debugging"
-	depends on DEBUG_KERNEL && HIGHMEM
+	depends on DEBUG_KERNEL && HIGHMEM && !IPIPE
 	help
 	  This option enables additional error checking for high memory
 	  systems.  Disable for production systems.
@@ -746,7 +751,7 @@ config ARCH_HAS_KCOV
 
 config KCOV
 	bool "Code coverage for fuzzing"
-	depends on ARCH_HAS_KCOV
+	depends on ARCH_HAS_KCOV && !IPIPE
 	select DEBUG_FS
 	select GCC_PLUGINS if !COMPILE_TEST
 	select GCC_PLUGIN_SANCOV if !COMPILE_TEST
@@ -787,7 +792,7 @@ config LOCKUP_DETECTOR
 
 config SOFTLOCKUP_DETECTOR
 	bool "Detect Soft Lockups"
-	depends on DEBUG_KERNEL && !S390
+	depends on DEBUG_KERNEL && !S390 && !IPIPE
 	select LOCKUP_DETECTOR
 	help
 	  Say Y here to enable the kernel to act as a watchdog to detect
@@ -815,7 +820,7 @@ config HARDLOCKUP_CHECK_TIMESTAMP
 #
 config HARDLOCKUP_DETECTOR
 	bool "Detect Hard Lockups"
-	depends on DEBUG_KERNEL && !S390
+	depends on DEBUG_KERNEL && !S390 && !IPIPE
 	depends on HAVE_HARDLOCKUP_DETECTOR_PERF || HAVE_HARDLOCKUP_DETECTOR_ARCH
 	select LOCKUP_DETECTOR
 	select HARDLOCKUP_DETECTOR_PERF if HAVE_HARDLOCKUP_DETECTOR_PERF
@@ -873,7 +878,7 @@ config BOOTPARAM_SOFTLOCKUP_PANIC_VALUE
 
 config DETECT_HUNG_TASK
 	bool "Detect Hung Tasks"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	default SOFTLOCKUP_DETECTOR
 	help
 	  Say Y here to enable the kernel to detect "hung tasks",
@@ -927,7 +932,7 @@ config BOOTPARAM_HUNG_TASK_PANIC_VALUE
 
 config WQ_WATCHDOG
 	bool "Detect Workqueue Stalls"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  Say Y here to enable stall detection on workqueues.  If a
 	  worker pool doesn't make forward progress on a pending work
@@ -940,6 +945,7 @@ endmenu # "Debug lockups and hangs"
 
 config PANIC_ON_OOPS
 	bool "Panic on Oops"
+	default y
 	help
 	  Say Y here to enable the kernel to panic when it oopses. This
 	  has the same effect as setting oops=panic on the kernel command
@@ -949,7 +955,7 @@ config PANIC_ON_OOPS
 	  anything erroneous after an oops which could result in data
 	  corruption or other issues.
 
-	  Say N if unsure.
+	  Say Y if unsure.
 
 config PANIC_ON_OOPS_VALUE
 	int
@@ -968,7 +974,7 @@ config PANIC_TIMEOUT
 
 config SCHED_DEBUG
 	bool "Collect scheduler debugging info"
-	depends on DEBUG_KERNEL && PROC_FS
+	depends on DEBUG_KERNEL && PROC_FS && !IPIPE
 	default y
 	help
 	  If you say Y here, the /proc/sched_debug file will be provided
@@ -981,7 +987,7 @@ config SCHED_INFO
 
 config SCHEDSTATS
 	bool "Collect scheduler statistics"
-	depends on DEBUG_KERNEL && PROC_FS
+	depends on DEBUG_KERNEL && PROC_FS && !IPIPE
 	select SCHED_INFO
 	help
 	  If you say Y here, additional code will be inserted into the
@@ -994,7 +1000,7 @@ config SCHEDSTATS
 
 config SCHED_STACK_END_CHECK
 	bool "Detect stack corruption on calls to schedule()"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	default n
 	help
 	  This option checks for a stack overrun on calls to schedule().
@@ -1006,6 +1012,7 @@ config SCHED_STACK_END_CHECK
 
 config DEBUG_TIMEKEEPING
 	bool "Enable extra timekeeping sanity checking"
+	depends on !IPIPE
 	help
 	  This option will enable additional timekeeping sanity checks
 	  which may be helpful when diagnosing issues where timekeeping
@@ -1019,7 +1026,7 @@ config DEBUG_TIMEKEEPING
 
 config DEBUG_PREEMPT
 	bool "Debug preemptible kernel"
-	depends on DEBUG_KERNEL && PREEMPT && TRACE_IRQFLAGS_SUPPORT
+	depends on DEBUG_KERNEL && PREEMPT && TRACE_IRQFLAGS_SUPPORT && !IPIPE
 	default y
 	help
 	  If you say Y here then the kernel will use a debug variant of the
@@ -1031,14 +1038,14 @@ menu "Lock Debugging (spinlocks, mutexes, etc...)"
 
 config DEBUG_RT_MUTEXES
 	bool "RT Mutex debugging, deadlock detection"
-	depends on DEBUG_KERNEL && RT_MUTEXES
+	depends on DEBUG_KERNEL && RT_MUTEXES && !IPIPE
 	help
 	 This allows rt mutex semantics violations and rt mutex related
 	 deadlocks (lockups) to be detected and reported automatically.
 
 config DEBUG_SPINLOCK
 	bool "Spinlock and rw-lock debugging: basic checks"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	select UNINLINE_SPIN_UNLOCK
 	help
 	  Say Y here and build SMP to catch missing spinlock initialization
@@ -1048,7 +1055,7 @@ config DEBUG_SPINLOCK
 
 config DEBUG_MUTEXES
 	bool "Mutex debugging: basic checks"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	 This feature allows mutex semantics violations to be detected and
 	 reported.
@@ -1056,6 +1063,7 @@ config DEBUG_MUTEXES
 config DEBUG_WW_MUTEX_SLOWPATH
 	bool "Wait/wound mutex debugging: Slowpath testing"
 	depends on DEBUG_KERNEL && TRACE_IRQFLAGS_SUPPORT && STACKTRACE_SUPPORT && LOCKDEP_SUPPORT
+	depends on !IPIPE
 	select DEBUG_LOCK_ALLOC
 	select DEBUG_SPINLOCK
 	select DEBUG_MUTEXES
@@ -1073,6 +1081,7 @@ config DEBUG_WW_MUTEX_SLOWPATH
 config DEBUG_LOCK_ALLOC
 	bool "Lock debugging: detect incorrect freeing of live locks"
 	depends on DEBUG_KERNEL && TRACE_IRQFLAGS_SUPPORT && STACKTRACE_SUPPORT && LOCKDEP_SUPPORT
+	depends on !IPIPE
 	select DEBUG_SPINLOCK
 	select DEBUG_MUTEXES
 	select DEBUG_RT_MUTEXES if RT_MUTEXES
@@ -1088,6 +1097,7 @@ config DEBUG_LOCK_ALLOC
 config PROVE_LOCKING
 	bool "Lock debugging: prove locking correctness"
 	depends on DEBUG_KERNEL && TRACE_IRQFLAGS_SUPPORT && STACKTRACE_SUPPORT && LOCKDEP_SUPPORT
+	depends on !IPIPE
 	select LOCKDEP
 	select DEBUG_SPINLOCK
 	select DEBUG_MUTEXES
@@ -1134,6 +1144,7 @@ config PROVE_LOCKING
 config LOCKDEP
 	bool
 	depends on DEBUG_KERNEL && TRACE_IRQFLAGS_SUPPORT && STACKTRACE_SUPPORT && LOCKDEP_SUPPORT
+	depends on !IPIPE
 	select STACKTRACE
 	select FRAME_POINTER if !MIPS && !PPC && !ARM_UNWIND && !S390 && !MICROBLAZE && !ARC && !SCORE && !X86
 	select KALLSYMS
@@ -1145,6 +1156,7 @@ config LOCKDEP_SMALL
 config LOCK_STAT
 	bool "Lock usage statistics"
 	depends on DEBUG_KERNEL && TRACE_IRQFLAGS_SUPPORT && STACKTRACE_SUPPORT && LOCKDEP_SUPPORT
+	depends on !IPIPE
 	select LOCKDEP
 	select DEBUG_SPINLOCK
 	select DEBUG_MUTEXES
@@ -1191,7 +1203,7 @@ config DEBUG_LOCKDEP
 config DEBUG_ATOMIC_SLEEP
 	bool "Sleep inside atomic section checking"
 	select PREEMPT_COUNT
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  If you say Y here, various routines which may sleep will become very
 	  noisy if they are called inside atomic sections: when a spinlock is
@@ -1200,7 +1212,7 @@ config DEBUG_ATOMIC_SLEEP
 
 config DEBUG_LOCKING_API_SELFTESTS
 	bool "Locking API boot-time self-tests"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  Say Y here if you want the kernel to run a short self-test during
 	  bootup. The self-test checks whether common types of locking bugs
@@ -1211,7 +1223,7 @@ config DEBUG_LOCKING_API_SELFTESTS
 
 config LOCK_TORTURE_TEST
 	tristate "torture tests for locking"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	select TORTURE_TEST
 	default n
 	help
@@ -1226,6 +1238,7 @@ config LOCK_TORTURE_TEST
 
 config WW_MUTEX_SELFTEST
 	tristate "Wait/wound mutex selftests"
+	depends on !IPIPE
 	help
 	  This option provides a kernel module that runs tests on the
 	  on the struct ww_mutex locking API.
@@ -1283,7 +1296,7 @@ config WARN_ALL_UNSEEDED_RANDOM
 
 config DEBUG_KOBJECT
 	bool "kobject debugging"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  If you say Y here, some extra kobject debugging messages will be sent
 	  to the syslog. 
@@ -1330,7 +1343,7 @@ config DEBUG_LIST
 
 config DEBUG_PI_LIST
 	bool "Debug priority linked list manipulation"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  Enable this to turn on extended checks in the priority-ordered
 	  linked-list (plist) walking routines.  This checks the entire
@@ -1340,7 +1353,7 @@ config DEBUG_PI_LIST
 
 config DEBUG_SG
 	bool "Debug SG table operations"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  Enable this to turn on checks on scatter-gather tables. This can
 	  help find problems with drivers that do not properly initialize
@@ -1350,7 +1363,7 @@ config DEBUG_SG
 
 config DEBUG_NOTIFIERS
 	bool "Debug notifier call chains"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  Enable this to turn on sanity checking for notifier call chains.
 	  This is most useful for kernel developers to make sure that
@@ -1360,7 +1373,7 @@ config DEBUG_NOTIFIERS
 
 config DEBUG_CREDENTIALS
 	bool "Debug credential management"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  Enable this to turn on some debug checking for credential
 	  management.  The additional code keeps track of the number of
@@ -1377,7 +1390,7 @@ source "kernel/rcu/Kconfig.debug"
 
 config DEBUG_WQ_FORCE_RR_CPU
 	bool "Force round-robin CPU selection for unbound work items"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	default n
 	help
 	  Workqueue used to implicitly guarantee that work items queued
@@ -1392,8 +1405,7 @@ config DEBUG_WQ_FORCE_RR_CPU
 
 config DEBUG_BLOCK_EXT_DEVT
         bool "Force extended block device numbers and spread them"
-	depends on DEBUG_KERNEL
-	depends on BLOCK
+	depends on DEBUG_KERNEL && BLOCK && !IPIPE
 	default n
 	help
 	  BIG FAT WARNING: ENABLING THIS OPTION MIGHT BREAK BOOTING ON
@@ -1432,7 +1444,7 @@ config CPU_HOTPLUG_STATE_CONTROL
 
 config NOTIFIER_ERROR_INJECTION
 	tristate "Notifier error injection"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	select DEBUG_FS
 	help
 	  This option provides the ability to inject artificial errors to
@@ -1507,7 +1519,7 @@ config NETDEV_NOTIFIER_ERROR_INJECT
 
 config FAULT_INJECTION
 	bool "Fault-injection framework"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  Provide fault-injection framework.
 	  For more details, see Documentation/fault-injection/.
@@ -1576,9 +1588,7 @@ config FAULT_INJECTION_STACKTRACE_FILTER
 
 config LATENCYTOP
 	bool "Latency measuring infrastructure"
-	depends on DEBUG_KERNEL
-	depends on STACKTRACE_SUPPORT
-	depends on PROC_FS
+	depends on DEBUG_KERNEL && STACKTRACE_SUPPORT && PROC_FS && !IPIPE
 	select FRAME_POINTER if !MIPS && !PPC && !S390 && !MICROBLAZE && !ARM_UNWIND && !ARC && !X86
 	select KALLSYMS
 	select KALLSYMS_ALL
@@ -1593,7 +1603,7 @@ source kernel/trace/Kconfig
 
 config PROVIDE_OHCI1394_DMA_INIT
 	bool "Remote debugging over FireWire early on boot"
-	depends on PCI && X86
+	depends on PCI && X86 && !IPIPE
 	help
 	  If you want to debug problems which hang or crash the kernel early
 	  on boot and the crashing machine has a FireWire port, you can use
@@ -1622,7 +1632,7 @@ config PROVIDE_OHCI1394_DMA_INIT
 
 config DMA_API_DEBUG
 	bool "Enable debugging of DMA-API usage"
-	depends on HAVE_DMA_API_DEBUG
+	depends on HAVE_DMA_API_DEBUG && !IPIPE
 	help
 	  Enable this option to debug the use of the DMA API by device drivers.
 	  With this option you will be able to detect common bugs in device
@@ -1643,8 +1653,7 @@ menu "Runtime Testing"
 
 config LKDTM
 	tristate "Linux Kernel Dump Test Tool Module"
-	depends on DEBUG_FS
-	depends on BLOCK
+	depends on DEBUG_FS && BLOCK && !IPIPE
 	default n
 	help
 	This module enables testing of the different dumping mechanisms by
@@ -1659,6 +1668,7 @@ config LKDTM
 config TEST_LIST_SORT
 	tristate "Linked list sorting test"
 	depends on DEBUG_KERNEL || m
+	depends on !IPIPE
 	help
 	  Enable this to turn on 'list_sort()' function test. This test is
 	  executed only once during system boot (so affects only boot time),
@@ -1669,6 +1679,7 @@ config TEST_LIST_SORT
 config TEST_SORT
 	tristate "Array-based sort test"
 	depends on DEBUG_KERNEL || m
+	depends on !IPIPE
 	help
 	  This option enables the self-test function of 'sort()' at boot,
 	  or at module load time.
@@ -1677,8 +1688,7 @@ config TEST_SORT
 
 config KPROBES_SANITY_TEST
 	bool "Kprobes sanity tests"
-	depends on DEBUG_KERNEL
-	depends on KPROBES
+	depends on DEBUG_KERNEL && KPROBES && !IPIPE
 	default n
 	help
 	  This option provides for testing basic kprobes functionality on
@@ -1689,7 +1699,7 @@ config KPROBES_SANITY_TEST
 
 config BACKTRACE_SELF_TEST
 	tristate "Self test for the backtrace code"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	default n
 	help
 	  This option provides a kernel module that can be used to test
@@ -1704,21 +1714,21 @@ config BACKTRACE_SELF_TEST
 
 config RBTREE_TEST
 	tristate "Red-Black tree test"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  A benchmark measuring the performance of the rbtree library.
 	  Also includes rbtree invariant checks.
 
 config INTERVAL_TREE_TEST
 	tristate "Interval tree test"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	select INTERVAL_TREE
 	help
 	  A benchmark measuring the performance of the interval tree library
 
 config PERCPU_TEST
 	tristate "Per cpu operations test"
-	depends on m && DEBUG_KERNEL
+	depends on m && DEBUG_KERNEL && !IPIPE
 	help
 	  Enable this option to build test module which validates per-cpu
 	  operations.
@@ -1727,6 +1737,7 @@ config PERCPU_TEST
 
 config ATOMIC64_SELFTEST
 	tristate "Perform an atomic64_t self-test"
+	depends on !IPIPE
 	help
 	  Enable this option to test the atomic64_t functions at boot or
 	  at module load time.
@@ -1735,7 +1746,7 @@ config ATOMIC64_SELFTEST
 
 config ASYNC_RAID6_TEST
 	tristate "Self test for hardware accelerated raid6 recovery"
-	depends on ASYNC_RAID6_RECOV
+	depends on ASYNC_RAID6_RECOV && !IPIPE
 	select ASYNC_MEMCPY
 	---help---
 	  This is a one-shot self test that permutes through the
@@ -1748,18 +1759,23 @@ config ASYNC_RAID6_TEST
 
 config TEST_HEXDUMP
 	tristate "Test functions located in the hexdump module at runtime"
+	depends on !IPIPE
 
 config TEST_STRING_HELPERS
 	tristate "Test functions located in the string_helpers module at runtime"
+	depends on !IPIPE
 
 config TEST_KSTRTOX
 	tristate "Test kstrto*() family of functions at runtime"
+	depends on !IPIPE
 
 config TEST_PRINTF
 	tristate "Test printf() family of functions at runtime"
+	depends on !IPIPE
 
 config TEST_BITMAP
 	tristate "Test bitmap_*() family of functions at runtime"
+	depends on !IPIPE
 	default n
 	help
 	  Enable this option to test the bitmap functions at boot.
@@ -1768,9 +1784,11 @@ config TEST_BITMAP
 
 config TEST_UUID
 	tristate "Test functions located in the uuid module at runtime"
+	depends on !IPIPE
 
 config TEST_RHASHTABLE
 	tristate "Perform selftest on resizable hash table"
+	depends on !IPIPE
 	default n
 	help
 	  Enable this option to test the rhashtable functions at boot.
@@ -1779,6 +1797,7 @@ config TEST_RHASHTABLE
 
 config TEST_HASH
 	tristate "Perform selftest on hash functions"
+	depends on !IPIPE
 	default n
 	help
 	  Enable this option to test the kernel's integer (<linux/hash.h>),
@@ -1791,7 +1810,7 @@ config TEST_HASH
 config TEST_PARMAN
 	tristate "Perform selftest on priority array manager"
 	default n
-	depends on PARMAN
+	depends on PARMAN && !IPIPE
 	help
 	  Enable this option to test priority array manager on boot
 	  (or module load).
@@ -1801,7 +1820,7 @@ config TEST_PARMAN
 config TEST_LKM
 	tristate "Test module loading with 'hello world' module"
 	default n
-	depends on m
+	depends on m && !IPIPE
 	help
 	  This builds the "test_module" module that emits "Hello, world"
 	  on printk when loaded. It is designed to be used for basic
@@ -1815,7 +1834,7 @@ config TEST_LKM
 config TEST_USER_COPY
 	tristate "Test user/kernel boundary protections"
 	default n
-	depends on m
+	depends on m && !IPIPE
 	help
 	  This builds the "test_user_copy" module that runs sanity checks
 	  on the copy_to/from_user infrastructure, making sure basic
@@ -1828,7 +1847,7 @@ config TEST_USER_COPY
 config TEST_BPF
 	tristate "Test BPF filter functionality"
 	default n
-	depends on m && NET
+	depends on m && NET && !IPIPE
 	help
 	  This builds the "test_bpf" module that runs various test vectors
 	  against the BPF interpreter or BPF JIT compiler depending on the
@@ -1842,7 +1861,7 @@ config TEST_BPF
 config TEST_FIRMWARE
 	tristate "Test firmware loading via userspace interface"
 	default n
-	depends on FW_LOADER
+	depends on FW_LOADER && !IPIPE
 	help
 	  This builds the "test_firmware" module that creates a userspace
 	  interface for testing firmware loading. This can be used to
@@ -1855,7 +1874,7 @@ config TEST_FIRMWARE
 config TEST_SYSCTL
 	tristate "sysctl test driver"
 	default n
-	depends on PROC_SYSCTL
+	depends on PROC_SYSCTL && !IPIPE
 	help
 	  This builds the "test_sysctl" module. This driver enables to test the
 	  proc sysctl interfaces available to drivers safely without affecting
@@ -1866,6 +1885,7 @@ config TEST_SYSCTL
 config TEST_UDELAY
 	tristate "udelay test driver"
 	default n
+	depends on !IPIPE
 	help
 	  This builds the "udelay_test" module that helps to make sure
 	  that udelay() is working properly.
@@ -1875,7 +1895,7 @@ config TEST_UDELAY
 config TEST_STATIC_KEYS
 	tristate "Test static keys"
 	default n
-	depends on m
+	depends on m && !IPIPE
 	help
 	  Test the static key interfaces.
 
@@ -1884,7 +1904,7 @@ config TEST_STATIC_KEYS
 config TEST_KMOD
 	tristate "kmod stress tester"
 	default n
-	depends on m
+	depends on m && !IPIPE
 	depends on BLOCK && (64BIT || LBDAF)	  # for XFS, BTRFS
 	depends on NETDEVICES && NET_CORE && INET # for TUN
 	depends on BLOCK
@@ -1923,7 +1943,7 @@ endmenu # runtime tests
 
 config MEMTEST
 	bool "Memtest"
-	depends on HAVE_MEMBLOCK
+	depends on HAVE_MEMBLOCK && !IPIPE
 	---help---
 	  This option adds a kernel parameter 'memtest', which allows memtest
 	  to be set.
diff --git a/lib/Kconfig.kasan b/lib/Kconfig.kasan
index 3d35d062970d..507adf6d287b 100644
--- a/lib/Kconfig.kasan
+++ b/lib/Kconfig.kasan
@@ -6,6 +6,7 @@ if HAVE_ARCH_KASAN
 config KASAN
 	bool "KASan: runtime memory debugger"
 	depends on SLUB || (SLAB && !DEBUG_SLAB)
+	depends on !IPIPE
 	select CONSTRUCTORS
 	select STACKDEPOT
 	help
diff --git a/lib/Kconfig.kgdb b/lib/Kconfig.kgdb
index ab4ff0eea776..b12066b1a691 100644
--- a/lib/Kconfig.kgdb
+++ b/lib/Kconfig.kgdb
@@ -5,7 +5,7 @@ config HAVE_ARCH_KGDB
 menuconfig KGDB
 	bool "KGDB: kernel debugger"
 	depends on HAVE_ARCH_KGDB
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	help
 	  If you say Y here, it will be possible to remotely debug the
 	  kernel using gdb.  It is recommended but not required, that
diff --git a/lib/Kconfig.ubsan b/lib/Kconfig.ubsan
index a669c193b878..17a668f8aefd 100644
--- a/lib/Kconfig.ubsan
+++ b/lib/Kconfig.ubsan
@@ -6,6 +6,7 @@ config ARCH_WANTS_UBSAN_NO_NULL
 
 config UBSAN
 	bool "Undefined behaviour sanity checker"
+	depends on !IPIPE
 	help
 	  This option enables undefined behaviour sanity checker
 	  Compile-time instrumentation is used to detect various undefined
diff --git a/mm/Kconfig b/mm/Kconfig
index cc4d633947bf..a5641711e929 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -160,7 +160,7 @@ config HAVE_BOOTMEM_INFO_NODE
 config MEMORY_HOTPLUG
 	bool "Allow for memory hot-add"
 	depends on SPARSEMEM || X86_64_ACPI_NUMA
-	depends on ARCH_ENABLE_MEMORY_HOTPLUG
+	depends on ARCH_ENABLE_MEMORY_HOTPLUG && !IPIPE
 
 config MEMORY_HOTPLUG_SPARSE
 	def_bool y
diff --git a/mm/Kconfig.debug b/mm/Kconfig.debug
index e5e606ee5f71..1368ae779904 100644
--- a/mm/Kconfig.debug
+++ b/mm/Kconfig.debug
@@ -1,5 +1,6 @@
 config PAGE_EXTENSION
 	bool "Extend memmap on extra space for more information on page"
+	depends on !IPIPE
 	---help---
 	  Extend memmap on extra space for more information on page. This
 	  could be used for debugging features that need to insert extra
@@ -9,7 +10,7 @@ config PAGE_EXTENSION
 
 config DEBUG_PAGEALLOC
 	bool "Debug page memory allocations"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	depends on !HIBERNATION || ARCH_SUPPORTS_DEBUG_PAGEALLOC && !PPC && !SPARC
 	select PAGE_EXTENSION
 	select PAGE_POISONING if !ARCH_SUPPORTS_DEBUG_PAGEALLOC
@@ -41,6 +42,7 @@ config DEBUG_PAGEALLOC_ENABLE_DEFAULT
 
 config PAGE_POISONING
 	bool "Poison pages after freeing"
+	depends on !IPIPE
 	select PAGE_POISONING_NO_SANITY if HIBERNATION
 	---help---
 	  Fill the pages with poison patterns after free_pages() and verify
@@ -79,7 +81,7 @@ config PAGE_POISONING_ZERO
 
 config DEBUG_PAGE_REF
 	bool "Enable tracepoint to track down page reference manipulation"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL && !IPIPE
 	depends on TRACEPOINTS
 	---help---
 	  This is a feature to add tracepoint for tracking down page reference
diff --git a/samples/Kconfig b/samples/Kconfig
index 9cb63188d3ef..4c4c6a07d15c 100644
--- a/samples/Kconfig
+++ b/samples/Kconfig
@@ -1,5 +1,6 @@
 menuconfig SAMPLES
 	bool "Sample kernel code"
+	depends on !IPIPE
 	help
 	  You can build and test sample kernel code here.
 
diff --git a/security/Kconfig b/security/Kconfig
index 87f2a6f842fd..1dd3ec658282 100644
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -22,6 +22,7 @@ config SECURITY
 	bool "Enable different security models"
 	depends on SYSFS
 	depends on MULTIUSER
+	depends on !IPIPE
 	help
 	  This allows you to choose different security modules to be
 	  configured into your kernel.
@@ -38,6 +39,7 @@ config SECURITY_WRITABLE_HOOKS
 
 config SECURITYFS
 	bool "Enable the securityfs filesystem"
+	depends on !IPIPE
 	help
 	  This will build the securityfs filesystem.  It is currently used by
 	  the TPM bios character driver and IMA, an integrity provider.  It is
@@ -56,7 +58,7 @@ config SECURITY_NETWORK
 
 config PAGE_TABLE_ISOLATION
 	bool "Remove the kernel mapping in user mode"
-	depends on X86_64 && !UML
+	depends on X86_64 && !UML && !IPIPE
 	default y
 	help
 	  This feature reduces the number of hardware side channels by
@@ -152,7 +154,7 @@ config HAVE_HARDENED_USERCOPY_ALLOCATOR
 
 config HARDENED_USERCOPY
 	bool "Harden memory copies between kernel and userspace"
-	depends on HAVE_HARDENED_USERCOPY_ALLOCATOR
+	depends on HAVE_HARDENED_USERCOPY_ALLOCATOR && !IPIPE
 	select BUG
 	imply STRICT_DEVMEM
 	help
-- 
2.22.0

